# <center>编译式RAG第二节课：GBrain 核心功能详解</center>

&emsp;&emsp;**先交代 GBrain 的来历**——它不是凭空冒出来的工具，而是一个有明确思想源头的范式落地。2026 年 4 月，Andrej Karpathy（前 OpenAI 创始成员、前 Tesla AI 总监，业内常称"K 神"）发布了一份名为 **LLM Wiki** 的"想法文件"（idea file）——一篇只有概念、不含代码的 GitHub Gist。它提出一个朴素却锋利的问题：我们用 RAG 处理文档时，大模型**每次回答都从零开始重新检索、重新拼凑，什么都没沉淀下来**；那能不能换个思路，让大模型像维护一个 wiki 一样，把读过的资料**一次性"编译"成结构化、互相链接的 markdown 知识库**，往后查询直接命中这份不断复利积累的产物？这就是 **LLM Wiki** 的核心主张，也是我们整个系列所称"**编译式 RAG**"的思想原点。（这一思路在精神上还能上溯到 1945 年 Vannevar Bush 设想的 Memex——只是 Bush 没解决"谁来做维护"，而大模型恰好把这件苦差事的成本压到了趋近于零。）

&emsp;&emsp;但 Karpathy 这份文件**有意保持抽象**，原文明说"描述的是思路、不是某个具体实现"——确切的目录结构、schema、工具选择都留给读者自己和 Agent 去落地。真正把它从一页概念变成**可安装、可跑、有 benchmark 的生产级系统**的，是 **Garry Tan** 开源的 **GBrain**（仓库 https://github.com/garrytan/gbrain）。GBrain 不止是照搬，还在 Karpathy 原版之上往前走了两步：一是**零 LLM 的 self-wiring 自动建图**，二是真正**替你把多页综合成带引用的答案、并诚实标出知识空白**的综合层；作者本人就用它跑着一个十几万页规模、几十个 cron job 夜间自动维护的真实大脑——这不是 Demo。**GBrain 正是这门课要讲透的主角**：借它，我们把 Karpathy 那页"想法文件"亲手落到能运行的代码上。

&emsp;&emsp;在进入正题前，先用一分钟交代背景——没接触过前置概念也能跟上。我们这个系列讲的是 **编译式 RAG**：一种和传统向量 RAG 不同的知识管理范式。传统 RAG 把资料切成 chunk、算 embedding、写进机器索引，查询时临时召回片段、让大模型当场综合；编译式 RAG 则先把原始资料**一次性"编译"成一份人类可读、互相链接的 wiki**，查询时直接读这份已经提炼好的产物。区别不在"有没有中间产物"，而在它是**黑箱机器索引**还是**可读、可 diff、可审计的 wiki**。**GBrain** 就是这套范式的一个生产级实现。

> 📌 **30 秒骨架速览（没上过前置课也能跟上）**：编译式 RAG 的骨架是"三层 + 三操作"。**三层**按"谁能改什么"划分——`raw/`（原料层，人类拥有、只读，事实源头）、`wiki/`（知识层，LLM 拥有并持续维护的编译产物）、`CLAUDE.md`（规则层，人类写的 schema，约束 LLM 怎么编译）。**三操作**——`Ingest`（编译：读 raw 新原料，在 wiki 生成互链页，并更新导航 index.md 与日志 log.md）、`Query`（查询：从 index 导航、只读 wiki 不重读 raw，给带引用的答案）、`Lint`（巡检：查断链 / 孤儿页 / 过期内容）。GBrain 把这套范式工程化落地，并额外提供一项零 LLM 的 **self-wiring 自动建图**能力（第 1 章先讲它）。

&emsp;&emsp;回到 GBrain：到目前为止，我们用它做的事更像一台"建图工具"——扫描 markdown、抽取 wikilink、靠 self-wiring 把孤立笔记织成一张知识图谱，而且全程加着 `--no-embedding`、一个嵌入向量都没算过。这件事本身已经很漂亮，但它只用到了 GBrain 的一半能力。

&emsp;&emsp;另一半是什么？是当你深夜想问"明天见 Alice 之前，我到底该知道些什么"时，它不只甩给你五个相关页面让你自己读，而是替你读完这几页、跨页综合成一段带逐条引用的答案，最后还诚实地告诉你"这些信息里有哪些是过期的、哪些是自相矛盾的"。从"给你页面"到"给你答案"，这一步跨越，正是这节课要补全的另一半。

&emsp;&emsp;要走完这另一半，我们得把前面只顾着建图、刻意跳过的两层机制摊开看清楚：知识到底以什么形态存进了数据库（Storage 层），查询时它又是怎么从库里把对的页面捞出来排好序的（Retrieval 层）。把这两层铺好，综合层的 `think` 才有立足之地。这就是今天的主线。

> 📌 **目标受众与前置要求**：本课面向有 RAG 使用经验（用过或听说过向量库）的技术人——编译式 RAG 范式、三层三操作、self-wiring 自动建图与 wikilink 语法这些前置概念，上面的速览和第 1 章会简要交代，**没上过前一讲也能跟上**。技术上你需要一台能装 Bun 的机器和一个可用的 embedding API Key，**不需要**自己实现任何检索算法。

> 📌 **学完本节你将带走 6 件产物**：① 看懂 git markdown 与数据库的"源真相 vs 编译产物"双轨关系，能回答"改了源数据库会自动更新吗"；② 能区分 PGLite 与 Postgres 两引擎的适用场景，避开 `npm install -g gbrain` 抢注坑装上真 GBrain；③ 把前面跳过的 embedding 开起来，让知识带上语义向量；④ 讲清 HNSW + BM25 + RRF 三路混合检索为什么不能直接加分；⑤ 用同一句话对照 `search` 与 `think`，亲眼看到"页面列表"变成"带引用的综合答案 + 空白分析"；⑥ 区分 LongMemEval 与 BrainBench 两个 benchmark 的可信度，不把它们混在一起引用。

> 💡 **学完不能做（诚实划界）**：本节聚焦"把 GBrain 当作完整 brain layer 讲透"。8.3 节给了自定义 schema 的最小入口和方法论，但完整的领域 schema 建模与效果验证、把 GBrain 接进 Claude Code 当 Agent 跨会话长记忆、团队多人隔离这些落地实战，本节先把机制和判断力打牢、不展开。

> 📅 **时效性说明**：本课全部命令与输出基于 GBrain v0.42.x 系列（实验室实测版本 0.42.44.0，仓库 https://github.com/garrytan/gbrain）。GBrain 处在活跃迭代期，小版本之间会有 breaking change——这一点本身就是一个诚实的能力边界，所以我们锁的是系列号、不锁死小版本。benchmark 数字引自官方评测仓库 gbrain-evals 的公开发布，每个数字都会标清来源与口径。

---

## <center>第 1 章：回顾与本节目标</center>

&emsp;&emsp;开始补全之前，我们先花一分钟认识 GBrain 已经会做的那一半——建图。这一章做两件事：先把 GBrain 的第一项能力（self-wiring 自动建图）简要讲清，再把"它已经能做的这半边"和"今天要补的另一半"并排摆出来，让缺口自己显现。看清了缺口，后面五站路才知道为什么要走。

### 1.1 self-wiring 自动建图

&emsp;&emsp;先用半分钟认识 GBrain 已经会做的第一件事——把一批 markdown 笔记**自动织成一张知识图谱**，而且全程不调用任何 LLM。它的做法叫 **self-wiring（自动接线）**：你在 markdown 里用 `[[companies/nexaflow]]` 这种**带目录前缀**的 wikilink 语法，显式写下页面之间的链接，GBrain 就用一组正则规则把这些链接解析成**带类型的有向边**（typed edges），图谱自己就"接"出来了。

&emsp;&emsp;举个例子：`alice-chen` 这页里写了 "founder of `[[companies/nexaflow]]`"，GBrain 解析后就自动生成一条 `alice-chen --founded--> nexaflow` 的有向边；常见关系类型有 `founded`（创办）、`works_at`（任职）、`invested_in`（投资）、`advises`（顾问）。整个过程**零 LLM、确定性、可复现**——这跟 GraphRAG 那种"让大模型从散文里抽实体、推断关系"的路子是两条道：链接指向的**实体目标**是人显式写好的，**关系类型**则由一组确定性正则根据链接周围的上下文判定（上下文里出现 "founded" 就标 `founded` 边），全程不调 LLM；两者其实是为不同输入设计的（GraphRAG 啃没有链接的散文，GBrain 解析已写好链接的 markdown），不是简单的优劣之分。建图的命令实操本节先不展开，第 4 章会用第一讲那份 wiki 数据把它完整回放一遍；这里你只要记住一句话：**GBrain 的第一项本事，是把人写好的 wikilink 零成本地编译成一张可查询、可追溯的关系图谱**。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175016387.png" width=55%></div>

&emsp;&emsp;但这只是 GBrain 的一半。上面这套建图，全程都没有计算**嵌入向量**（embedding）——也就是说，这台大脑此刻只会"连"，还不会"读和想"。它能告诉你"Alice 和 NexaFlow 之间有一条 `founded` 边"，但你要是问它"明天见 Alice 前我该知道什么"，它就无能为力了：没有 embedding，它做不了语义检索，更别说跨页综合。**这另一半能力，正是这节课要补的。**

### 1.2 另一半能力：从页面到答案

&emsp;&emsp;那另一半到底是什么？用一张对比表把两边能力摆清楚，缺口一眼可见。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174314285.png" width=55%></div>

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>GBrain 的两半能力对照</font></p>
<div class="center">

| 维度 | GBrain 的一半：建图（self-wiring） | 今天要补的另一半：综合 |
|------|---------------------|------------------------|
| 核心动作 | self-wiring 自动建图、抽 wikilink | 读多页、跨源综合、写带引用的答案 |
| 关键开关 | 全程 `--no-embedding` | 必须开 embedding（语义向量） |
| 你得到什么 | 一张知识图谱 + 一批页面 | 一段成文答案 + 逐条引用 + 空白分析 |
| 典型命令 | `gbrain init` / `import` / graph 查询 | `gbrain embed` / `search` / `think` |
| 底层依赖 | 文件扫描 + 图谱表 | pgvector 向量 + 混合检索 + LLM 合成 |
| 回答得了的问题 | "Alice 和谁有关系" | "见 Alice 前我该准备什么" |

</div>

&emsp;&emsp;看最后一行就够了：左边那台大脑回答得了"谁和谁有关系"，但回答不了"我该准备什么"。后者需要它读完所有相关页面、把散落的事实综合成一段话、还要标清每句话出自哪一页。这个跨越——我们后面会反复称之为"从 search 到 think 的跨越"——就是这节课的灵魂。

&emsp;&emsp;这里先记一笔：等第 7 章真正跑通 `think` 的时候，你会看到它输出的那段答案末尾，还有一个最容易被忽略、却最有价值的部分。先有个印象，到那一章自然会讲到。

### 1.3 本节路线图

&emsp;&emsp;补全"另一半"不是一步到位的，它有依赖顺序。正式补全之前，第 3 章会先用一节确保你装的是真 GBrain——最直觉的装法会装错；装对了地基，第 4 章我们先用第一讲的 wiki 数据回顾一遍 self-wiring 建图，再走五站补全，每一站补上前面只建图、没碰的一块：

&emsp;&emsp;**第一站 · 架构全景**（第 2 章）：先建一张"brain 和 source 各管什么"的心智地图，避免后面把数据库和 git 仓库搞混。

&emsp;&emsp;**第二站 · Storage 层**（第 5 章）：补全前面建图时跳过的存储原理——PGLite 和 Postgres 两个引擎、第一次把 embedding 开起来。

&emsp;&emsp;**第三站 · Retrieval 层**（第 6 章）：补全前面完全没碰的混合检索——向量、关键词、RRF 三路怎么各补短板，再加两个工具坑实战。

&emsp;&emsp;**第四站 · 综合层**（第 7 章）：本节的重点，先看 `search` 的局限，再看 `think` 如何替你把多页综合成带引用的答案，核心目标在这里落地。

&emsp;&emsp;**第五站 · 生态与评判**（第 8、9 章）：51 个 skills 和约 90 个 MCP 工具是两回事、schema 包机制、夜间维护的 Minions 队列，最后用两个 benchmark 把"这东西到底行不行"的判断框架立起来。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174308665.png" width=65%></div>

&emsp;&emsp;走完五站，第 10 章我们盘点你能做什么、诚实标出本节没深入的边界。提醒一下阅读顺序：第一站"架构全景"就是紧接着的第 2 章，它是后面所有站的地图、要先建起来；之后第 3 章装机、第 4 章回顾建图打好地基，再从第 5 章起逐站补全。下面，从第 2 章架构全景开始。

---

## <center>第 2 章：整体架构（三层 + 两轴）</center>

&emsp;&emsp;真正动手补两层之前，必须先有一张地图。GBrain 里最容易让人犯迷糊的，是它同时有"数据库"和"git 仓库"两套东西，新手很容易把它们当成"主存储"和"备份"的关系——而真相恰好相反。这一章我们花点时间把这张地图画清楚：三层各管什么、brain 和 source 这两根轴怎么分工、谁才是"源真相"。地图对了，第 5、6 章深入两层时才不会迷路。

### 2.1 三层定位：存、连、取

&emsp;&emsp;GBrain 作为一个完整的 brain layer，内部分三层各司其职。（提醒：这里的"三层"指 GBrain 内部架构的存 / 连 / 取，和开头"30 秒骨架速览"里编译式 RAG 范式的"三层"——raw / wiki / 规则——是两个不同维度：一个讲 GBrain 怎么实现，一个讲编译式 RAG 怎么组织目录，别把两套"三层"搞混。）我们用一句话给每层定位：

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174312920.png" width=55%></div>

&emsp;&emsp;这张图把三层关系摆清楚了：

&emsp;&emsp;**Graph 层（连）**——这是我们刚在第 1 章认识的建图那层。它负责 self-wiring，把页面之间的关系抽成一张图谱。今天我们不重复讲它，但要记住它一直在那儿，第 7 章的 `think` 综合时会用到这些关系边。

&emsp;&emsp;**Storage 层（存）**——知识到底以什么形态存下来。这是第 5 章的主场：git 里的 markdown 是一种形态，数据库里的 chunk + 向量是另一种形态，两者什么关系，待会儿展开。

&emsp;&emsp;**Retrieval 层（取）**——查询时怎么把对的页面捞出来。这是第 6 章的主场：不是单跑一路检索，而是向量、关键词两路一起跑再融合。

&emsp;&emsp;今天我们主攻"存"和"取"这两层，因为它们正是前面建图时全程 `--no-embedding` 跳过的部分。"连"那层第 1 章讲过了，只在需要时点到。

### 2.2 两根轴：brain ⊥ source

&emsp;&emsp;三层之外，还有一个更基础的划分，理解它能省掉后面一大堆混淆——GBrain 把所有东西组织在两根正交的轴上：**brain（数据库）** 和 **source（仓库）**。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174228835.png" width=55%></div>

&emsp;&emsp;这两根轴的分工是：**source 是人维护的真相，brain 是从真相编译出来的派生物**。git 里的 markdown 你可以用编辑器改、用 `git diff` 看变更、用 git 历史回溯——它是唯一的、权威的源。而数据库里的 chunk 和向量，是 `gbrain import` 把 source 编译出来的产物，专门给机器做检索用。

&emsp;&emsp;为什么强调"正交"？因为它们职责完全不同、不可互相替代。你不会拿数据库当真相去人工读（它是切碎的 chunk + 一堆浮点数），也不会拿 markdown 去做向量检索（它没有向量）。两根轴各自完整，合起来才是一个能用的 brain layer。

### 2.3 源真相 vs 编译产物

&emsp;&emsp;这节课最该记住的一句话是：**git 里的 markdown 是源真相（source of truth），数据库里的 chunk + 向量是编译产物（compiled product）。** 这句话会贯穿整个第 5 章，值得停下来咀嚼一下它为什么重要。

&emsp;&emsp;这对关系和编译型编程语言里"源代码 vs 可执行文件"是同一个隐喻：源码是人维护的真相，可执行文件是编译出来给机器跑的产物。你改源码、重新编译、得到新的可执行文件——没人会反过来去手改可执行文件。GBrain 也一样：你改 markdown、重新 `import`、得到更新后的数据库。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174229733.png" width=55%></div>

> **【常见误区】**：很多人第一次接触会本能地把"数据库"当成主存储、把 git markdown 当成"备份"——毕竟数据库听起来更"正式"。这个直觉**正好反了**。后果是：你会去想办法直接改数据库里的内容，然后发现下次 `import` 时改动被覆盖、莫名其妙。正确的做法是：永远改 git 里的 markdown（真相），再 `gbrain import` 让数据库跟上。排查这类问题的方法很简单——任何时候对数据库内容有疑问，回去看对应的 markdown 源文件，它才是权威。

&emsp;&emsp;这句论断还引出一个具体的验收问题，我们会在第 5 章用一个动手实验把它跑出来：**改了 git 里的 markdown，数据库会自动更新吗？** 先把这个问题揣在兜里，到 5.1 节我们用真实数据回答它。

&emsp;&emsp;不过在深入各层动手之前，得先把最基础的一步落实——确认你机器上这套 `gbrain` 是真的、装对了。下面先过一遍装机。

---

## <center>第 3 章：安装与验证</center>

&emsp;&emsp;正式动手补两层之前，先落实最基础的一步：确认你机器上这套 `gbrain` 是真的、装对了。别小看这一步——`gbrain` 这个名字上踩着两个真实的坑，最直觉的两条安装路径都会装错，而装错了后面每一条命令都白跑。我们用"先用错、再选对"的方式走一遍。

### 3.1 三条安装路径对照

&emsp;&emsp;**第一步：最直觉的装法 `npm install -g gbrain`。** 装一个 CLI 工具，你的第一反应大概率是 npm。但对 GBrain 这是个陷阱。

&emsp;&emsp;下面这张实验截图把三条安装路径的真实结果放在同一个终端里对照，是避坑记忆的核心证据。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174303755.png" width=70%></div>

&emsp;&emsp;**这是抢注的 GPU 库，不是 GBrain。** npm 是开放注册表，任何人都能抢注包名。npm 上叫 `gbrain` 的包是 2018 年就被注册的一个 GPU 机器学习 JS 库（v1.3.1），描述写着 "GPU Javascript Library for Machine Learning"，和我们要的 GBrain 毫无关系。GBrain 从未在 npm 上发布过。装了它，你 `gbrain --version` 会看到 `1.3.1`——一眼就能识破，因为真 GBrain 是 `0.42.x`。

&emsp;&emsp;**第二步：换 `bun install -g github:garrytan/gbrain`。** 既然 GBrain 在 GitHub 上，直接从 GitHub 全局安装看似合理。但这条路也会断。

&emsp;&emsp;GBrain 的 `package.json` 带一个 `postinstall` 脚本，而 Bun 的全局安装模式出于安全考虑会对 postinstall 做沙箱限制，导致安装中断报 `error: postinstall script exited with code 1`。这不是 GBrain 的 bug，是 Bun 的安全机制与 GBrain 安装脚本的交互结果。

&emsp;&emsp;**第三步：正确路径——下载源码 + 本地 `bun install` + `bun link`。** 绕开上面两个坑的办法，是把源码拿到本地、做一次本地安装、再把 CLI 链接到全局 PATH。这条路径下 postinstall 在本地正常执行不受沙箱限制，装到的也是仓库里真正的 GBrain。完整命令拆开如下（注意这是真实的 shell 操作，放在 code cell 里用 `!` 前缀执行）。

In [ ]:
# 注意：以下为真实安装步骤的演示。已用注释保护，避免在本节环境误执行重装。
# 你在自己机器首次安装 GBrain 时，取消注释逐条运行。

# 1) 用 codeload CDN 下载仓库 ZIP（等价于 git clone；主域被墙时 CDN 节点通常可达）
# 注意：下载的是 master 主分支，可能比本课演示的 0.42.44.0 新；装完务必用 3.2 节 gbrain --version 核对版本，
#       差异大时部分命令/输出可能不同——这正是前面强调"锁版本"的原因（要严格复现可改用固定 tag）
# !curl --max-time 120 -s -L "https://codeload.github.com/garrytan/gbrain/zip/refs/heads/master" -o /tmp/gbrain-master.zip
# !curl --max-time 120 -s -L "https://codeload.github.com/garrytan/gbrain/zip/090bb53203557f5659563ea28c1c847c32167aeb" -o /tmp/gbrain-v0.42.44.0.zip #安装确定的版本

# 2) 解压到 ~/dev
# !mkdir -p ~/dev && unzip -o /tmp/gbrain-master.zip -d ~/dev/

# 3) 本地安装依赖（约 276 个包，含 PGLite WASM、pgvector）
# !cd ~/dev/gbrain-master && bun install

# 4) 把 gbrain CLI 链接到全局 PATH（链接落在 ~/.bun/bin/gbrain）
# !cd ~/dev/gbrain-master && bun link

print("安装步骤已用注释保护。首次安装时逐条取消注释运行。")

&emsp;&emsp;这几步里 `bun install` 在本地目录执行、postinstall 不受全局沙箱限制，正常装完依赖（含 PGLite 的 WASM 文件和 pgvector）。`bun link` 把当前目录的 `gbrain` 命令注册成全局软链接，落在 `~/.bun/bin/gbrain`，而这个目录已在 PATH 里，所以装完任意目录都能直接敲 `gbrain`。

### 3.2 验证装到的是真 GBrain

&emsp;&emsp;装完最关键的一步：验证装到的是真 GBrain，而不是那个抢注的 GPU 库。这是本章的 Tier 2 端到端验证。直接用全局命令 `gbrain --version`（这一步正是确认全局 PATH 里装的是哪个 `gbrain`）。

In [3]:
# Tier 2 端到端验证：确认装的是真 GBrain（版本号 0.42.x，而非抢注包的 1.3.1）
!gbrain --version
# 一眼识破：真 GBrain 是 0.42.x；如果看到 1.3.1，说明装成了 npm 抢注的 GPU 库

gbrain 0.42.44.0


&emsp;&emsp;输出 `gbrain 0.42.44.0` 就对了——这是真 GBrain。如果你这一步看到的是 `1.3.1`，立刻知道装错了，回到第三步走正确路径重装。把三条路径汇总成一张表，对照记忆最牢。

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>GBrain 三条安装路径对照</font></p>
<div class="center">

| 安装路径 | 结果 | 判断 |
|----------|------|------|
| `npm install -g gbrain` | 装到抢注的 GPU JS 库 v1.3.1 | 错（`--version` 显示 1.3.1） |
| `bun install -g github:garrytan/gbrain` | postinstall 被 Bun 沙箱拦截 | 错（安装中断） |
| 下载源码 + `bun install` + `bun link` | 装到真 GBrain v0.42.x | 对（`--version` 显示 0.42.x） |

</div>

> **【踩坑预警】**：为什么要锁版本？因为 GBrain 处在活跃迭代期，小版本之间会有 breaking change——后面讲 Storage 时你会撞到一个（`init` 的 embedding 行为在不同小版本不一样，见 5.2 节）、讲生态时还会撞到另一个（`schema detect` 直接报错，见 8.3 节）。后果是：照着一个版本的教程在另一个版本上跑，可能某条命令行为完全不同。正确做法是：装机时锁定一个具体小版本（实验室用的是 0.42.x 系列），团队内统一，避免"同一条命令在不同版本行为不同"的麻烦。这也是我们锁系列号、不锁死小版本数字的原因。

&emsp;&emsp;装对了 GBrain，地基就稳了。正式补全存储和检索之前，我们先用第一讲那份 wiki 数据，把建图命令真跑一遍——既确认你装的 GBrain 真能用，也回顾一下第一讲的 self-wiring。

---

## <center>第 4 章：self-wiring 建图回顾</center>

&emsp;&emsp;第一讲的第四章，我们学过 GBrain 的第一项本事——**self-wiring**：零 LLM、纯正则，把 markdown 里写好的 `[[wikilink]]` 自动接成一张知识图谱。那一讲全程 `--no-embedding`，只建图、不算向量。现在 GBrain 刚装好，正式补全存储和检索之前，我们用第一讲那份 wiki 数据把建图命令真跑一遍：一来确认环境可用，二来重温第一讲的手感，三来这张"只会连、还不会读"的图，正好是下一章要补的起点。

&emsp;&emsp;数据就是第一讲的 `llm-wiki-demo/wiki/`，5 个页面——`compiled-rag`（概念）、`gbrain`（系统）、`compiled-rag-cost`（成本争议）、`index`、`log`，彼此用 `[[...]]` 互链。

### 4.1 建库 + 导入：把 wiki 编译进大脑

&emsp;&emsp;和第一讲一样，这一步全程 `--no-embedding`——只把页面切块入库、建图，不算向量。先准备好两个路径变量。

In [4]:
# 回顾第一讲：用第一讲的 wiki 数据建图，全程 --no-embedding（只建图、不算向量）
# 第一讲的 wiki 语料目录（5 页，彼此用 [[wikilink]] 互链）
WIKI_DIR = "./llm-wiki-demo/wiki"
# 隔离到课件目录下的 output/ 子文件夹，产出就在课件旁边、清晰可见（不污染你已有的大脑）
import os
os.makedirs("output", exist_ok=True)
RECAP_PATH = "./output/brain-wiki-recap"
print("wiki 语料：  ", WIKI_DIR)
print("回顾用大脑：", RECAP_PATH)

wiki 语料：   ./llm-wiki-demo/wiki
回顾用大脑： ./output/brain-wiki-recap


&emsp;&emsp;建库时用 `--no-embedding`，明确告诉 GBrain"这个库不配向量模型"——这正是第一讲的用法，只建图、不算向量。

In [5]:
# 建库：--pglite 本地引擎，--no-embedding 不配向量模型（第一讲就是这么用的）
!gbrain init --pglite --no-embedding --force --path {RECAP_PATH} --non-interactive 2>&1 | tail -3


  Do NOT install without asking. The user owns this decision.


&emsp;&emsp;导入之前，先交代 GBrain 一个贯穿全课的关键机制：**GBrain 是「单活动库」模型**。`gbrain init` 不只是建库，它还把这个库登记成**当前活动库**（写进全局 `~/.gbrain/config.json` 的 `database_path`）；之后 `import`、`stats`、`graph`、`backlinks` 这些命令**默认就作用在活动库上，不需要每条都指定库路径**——事实上除了 `init` 等少数命令，其余命令即便你传了 `--path` 也会被直接忽略。所以本章接下来的命令你会看到都不带库参数，它们全部落在刚 `init` 出来的 `RECAP_PATH` 库上。**一个必须记住的推论**：等第 5 章再 `init` 一个新库，活动库就切走了；那时若想回头重跑本章某条命令，得先重新 `init` 回这个库，否则会跑到第 5 章的库上去。

&emsp;&emsp;库建好了，导入 wiki 的 5 个页面。因为库没配 embedding，导入时要加 `--no-embed` 跳过向量化——纯把 markdown 切成 chunk 写进库。

In [6]:
# 导入 wiki/ 5 页；--no-embed 跳过向量化（库本身也没配 embedding）
!gbrain import {WIKI_DIR} --no-embed

[gbrain phase] import.collect_files start dir=./llm-wiki-demo/wiki strategy=markdown
[gbrain phase] import.collect_files done 18ms files=5
Found 5 markdown files
[import.files] 5/5 (100%) donerted=5 skipped=0 errors=0

Import complete (0.1s):
  5 pages imported
  0 pages skipped (0 unchanged, 0 errors)
  8 chunks created


&emsp;&emsp;你会看到 `5 pages imported, 8 chunks created`——5 页全部入库，切出 8 个 chunk。这一步就是"写入时编译"：把人写的 markdown 编译成机器能检索的 chunk。但此刻**页面之间的边还没建**，下一节才是 self-wiring 的主角。

### 4.2 self-wiring 抽边：零 LLM 把 wikilink 接成图

&emsp;&emsp;页面进库了，但它们之间的链接关系还只是 markdown 文本里的 `[[...]]`，没有变成图谱的边。self-wiring 的核心命令是 `extract links`——它**纯正则解析每页的 `[[...]]`，确定性地建出有向边，全程零 LLM**。

In [7]:
# self-wiring 核心：extract links 纯正则解析 [[...]]，建有向边，零 LLM
# --source fs --dir：从文件系统目录直接抽（这份 wiki 用 bare [[slug]] 无前缀，--source db 建 0 边，fs 模式能处理）
!gbrain extract links --source fs --dir {WIKI_DIR}

[extract.links_fs] 3/3 (100%) done
Links: created 6 from 3 pages

Done: 6 links, 0 timeline entries from 3 pages


&emsp;&emsp;输出是 `Links: created 6 from 3 pages`——从 3 个有互链的页面（`compiled-rag` / `gbrain` / `compiled-rag-cost`）抽出 6 条边。这里有两个细节值得说清：

&emsp;&emsp;**为什么用 `--source fs --dir`，而不是 `--source db`？** `extract links` 有两种取源：`--source db` 从数据库里已导入的页抽，`--source fs --dir` 直接从文件系统目录抽。差别在对**链接格式**的处理：这份 wiki 用的是 **bare wikilink**（`[[compiled-rag]]`，不带目录前缀），实测 `--source db` 对它建 **0 条边**——db 模式认不出这种不带前缀的 bare slug；换成 `--source fs --dir` 直接扫文件就能正确处理，建出 6 条边。所以这里用 fs 模式。（如果你的链接带规范目录前缀、如 `[[people/alice]]`，两种模式都能抽。）

&emsp;&emsp;**为什么边的类型是 `mentions`？** self-wiring 的边类型，由一组**确定性正则**根据链接周围的上下文判定（零 LLM）：上下文里有 founded / invested in / advises / works at 这类动词，就标成对应的 `founded` / `invested_in` / `advises` / `works_at` 边；没有可识别的关系动词时，落到默认的 `mentions`（提及）。这份 wiki 的链接是 `concept` / `system` / `debate` 之间的泛指引用（index 里"概念 ↔ 系统 ↔ 争议"），上下文没有那些动词，所以全抽成 `mentions`。第一讲语料里"X founded Y"那种写法，才会得到 `founded` 这类 typed edge。

### 4.3 看图：可查询、可追溯

&emsp;&emsp;图建好了，用三个命令看看它长什么样。先看整体规模。

In [8]:
# stats：大脑规模总览——5 页、6 条边，全程没算向量（Embedded 0）
!gbrain stats 2>&1 | head -12

Pages:     5
Chunks:    8
Embedded:  0
Links:     6
Tags:      0
Timeline:  0

By type:
  note: 2
  debate: 1
  concept: 1
  system: 1


&emsp;&emsp;关键几行：`Pages 5` / `Links 6` / `Embedded 0`，按类型分 `note 2 / debate 1 / concept 1 / system 1`。**`Embedded 0` 正是"只建图、没算向量"的证据**——和第一讲完全一致。再从某一页出发，看它连到哪。

In [9]:
# graph：从 compiled-rag 这页出发，看它连到哪
!gbrain graph compiled-rag 2>&1 | head -20

[
  {
    "slug": "compiled-rag",
    "title": "编译式 RAG",
    "type": "concept",
    "depth": 0,
    "links": [
      {
        "to_slug": "compiled-rag-cost",
        "link_type": "mentions"
      },
      {
        "to_slug": "gbrain",
        "link_type": "mentions"
      }
    ]
  },
  {
    "slug": "compiled-rag-cost",
    "title": "编译式 RAG 的成本争议",


&emsp;&emsp;输出是一段 JSON 图谱：`compiled-rag`（concept）连到 `compiled-rag-cost` 和 `gbrain`，都是 `mentions` 边——和 index.md 里写的"概念 ↔ 系统 ↔ 争议"互链完全对上。最后反过来看，谁指向了某一页。

In [11]:
# backlinks：反向看，谁指向了 gbrain 这页
!gbrain backlinks gbrain 2>&1 | head -25

[
  {
    "from_slug": "compiled-rag-cost",
    "to_slug": "gbrain",
    "link_type": "mentions",
    "context": "markdown link: [系统]",
    "link_source": "markdown",
    "origin_slug": null,
    "origin_field": null
  },
  {
    "from_slug": "compiled-rag",
    "to_slug": "gbrain",
    "link_type": "mentions",
    "context": "markdown link: [gbrain]",
    "link_source": "markdown",
    "origin_slug": null,
    "origin_field": null
  }
]


&emsp;&emsp;输出显示 `compiled-rag` 和 `compiled-rag-cost` 都指向 `gbrain`，而且每条边都带 `context`（如 `markdown link: [gbrain]`）——**这就是 self-wiring "每条边都带可追溯的来源上下文" 的体现**——输出里能看到边来自哪一页、对应哪段 markdown 链接文本（记录的是上下文片段，不是精确行号），确定性、可审计，和 GraphRAG 那种 LLM 推断出来、说不清来源的边完全不同。

### 4.4 剖开产出物：这个 brain 到底是什么

&emsp;&emsp;建图、查图都跑通了，进下一站之前，我们把这一路产出的 brain 亲手剖开看一眼——它到底以什么形态存在磁盘上、能不能像普通数据库一样用 SQL 查。

&emsp;&emsp;**先看物理形态。** 我们这个 brain 就落在课件目录的 `./output/brain-wiki-recap`，打开它看一眼：

In [12]:
# brain 的物理落点：它不是一个 .db 单文件，而是一整套标准 PostgreSQL 数据目录
!ls {RECAP_PATH} && echo "--- PG_VERSION ---" && cat {RECAP_PATH}/PG_VERSION

PG_VERSION           pg_multixact         pg_tblspc
base                 pg_notify            pg_twophase
global               pg_replslot          pg_wal
pg_commit_ts         pg_serial            pg_xact
pg_dynshmem          pg_snapshots         postgresql.auto.conf
pg_hba.conf          pg_stat              postgresql.conf
pg_ident.conf        pg_stat_tmp
pg_logical           pg_subtrans
--- PG_VERSION ---
17


&emsp;&emsp;你会看到 `base/`（表数据，子目录是各 database 的 OID）、`global/`、`pg_wal/`、`postgresql.conf`、`PG_VERSION`（内容是 `17`）——全是真 Postgres 才有的东西。这把"编译产物 = 一个真数据库"落到了眼见为实：PGLite 不是某种轻量玩具格式，它就是**编译成 WASM 的 PostgreSQL 17**，数据在磁盘上跟你本地装的 Postgres 长得一模一样。

&emsp;&emsp;**再用真 SQL 查它的内部表。** 既然是 Postgres，能不能像查数据库一样 `SELECT` 它？能——但有个前提要先讲清：PGLite 是**进程内的 WASM Postgres、不监听任何端口**，所以你**没法用 `psql` 从外部连**它。正确方式是用 PGLite 的 JS 接口（在课件目录装一份与 GBrain 同版本 `0.4.3` 的 PGLite，自包含、不依赖 gbrain 装在哪）加载这个目录、跑 SQL：

In [13]:
# 用真 SQL 直接查 brain 内部表
# PGLite 是 WASM Postgres、不监听端口 → 不能 psql，只能用它的 JS 接口加载库目录
import subprocess, os

# 1) 一次性在课件目录装一份 PGLite（与 GBrain 同版本 0.4.3，含 vector/pg_trgm 扩展子导出）
#    自包含、不依赖 gbrain 装在哪；已装则秒过，node_modules / package.json 落在课件目录
subprocess.run(['bun', 'add', '@electric-sql/pglite@0.4.3'], capture_output=True, text=True)
brain_abs = os.path.abspath(RECAP_PATH)   # 直接指物理目录，精确可靠（绕过 gbrain 的活动库机制）

# 2) 一段 PGLite JS：加载库 + 跑 SQL（gbrain 建库时配了 vector/pg_trgm 扩展，加载必须带上，否则起不来）
js = r'''
import { PGlite } from '@electric-sql/pglite';
import { vector } from '@electric-sql/pglite/vector';
import { pg_trgm } from '@electric-sql/pglite/contrib/pg_trgm';
const db = await PGlite.create({ dataDir: process.argv[2], extensions: { vector, pg_trgm } });
const pages = await db.query('SELECT slug, type, title FROM pages ORDER BY type, slug');
console.log('[SQL] SELECT slug,type,title FROM pages  ->', pages.rows.length, 'rows');
for (const r of pages.rows) console.log('   ', r.slug.padEnd(20), '|', String(r.type).padEnd(8), '|', r.title);
const c = await db.query('SELECT count(*)::int n FROM content_chunks');
const l = await db.query('SELECT count(*)::int n FROM links');
console.log('[SQL] content_chunks =', c.rows[0].n, '| links =', l.rows[0].n);
await db.close();
'''
# 3) 脚本写在课件目录（bun 从这里解析刚装的本地 node_modules），跑完即删
mjs = '_inspect_brain_sql.mjs'
with open(mjs, 'w') as f: f.write(js)
try:
    r = subprocess.run(['bun', mjs, brain_abs], capture_output=True, text=True)
    print(r.stdout if r.stdout.strip() else r.stderr[:800])
finally:
    os.remove(mjs)

[SQL] SELECT slug,type,title FROM pages  -> 5 rows
    compiled-rag         | concept  | 编译式 RAG
    compiled-rag-cost    | debate   | 编译式 RAG 的成本争议
    index                | note     | Wiki 内容目录
    log                  | note     | Ingest 日志（append-only）
    gbrain               | system   | GBrain
[SQL] content_chunks = 8 | links = 6



&emsp;&emsp;输出会列出 `pages` 表里编译出的每一页（`compiled-rag` / `gbrain` / `index` / `log` 等，各带 `type` 和 `title`），以及 `content_chunks`（切块数）和 `links`（self-wiring 抽出的图谱边数）。这一眼就把编译式 RAG 的本质摊平了：**所谓"编译产物"，落到底就是几张结构化的 Postgres 表**——`pages` 存页面、`content_chunks` 存切块、`links` 存关系边。你前面用 `gbrain graph`、`gbrain stats`、`gbrain backlinks` 看到的一切，底层都是对这几张表的 SQL 查询；GBrain 的命令只是把 SQL 包了一层友好外壳。把这层壳掀开，你对"brain 到底是什么"就再没有黑箱了。

&emsp;&emsp;这张图全程没算一个向量——它只会"连"，还不会"读和想"。你能问它"`compiled-rag` 连到哪"，但你要是问"编译式 RAG 到底省不省成本"，它答不了：没有 embedding，做不了语义检索，更别说跨页综合。下一章先补前面跳过的存储原理（源真相、两个引擎），再把 embedding 开起来、让大脑带上语义坐标——那是 Storage 层最后一节的事。

---

## <center>第 5 章：Storage 存储层</center>

&emsp;&emsp;到了补全的第二站。前面建图时你 `init` 建库、`import` 导入，但库的内部长什么样、用的什么引擎、两轨怎么协同——这些全程没展开。这一章把它们补齐：先把"源真相 vs 编译产物"这对关系用真实实验坐实，再讲 PGLite 和 Postgres 两个存储引擎，最后第一次把前面跳过的 embedding 开起来。GBrain 已经装好、第一讲的图也在上一章回顾过了，这一章正式补上存储原理。

### 5.1 源真相 vs 编译产物：动手验证

&emsp;&emsp;第 2 章我们立了个论断"git 是源真相、数据库是编译产物"，并揣了个问题"改了源数据库会自动更新吗"。这一节就把它落到实处。先看一张图，把两轨的职责分离画清楚。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175043711.png" width=65%></div>

&emsp;&emsp;这两轨的具体体现是：git 里的 markdown 可读、可 `git diff`、可版本控制，是人维护的真相；数据库里同一页被切成 chunk、每个 chunk 配一个 1024 维向量，是机器检索用的派生物。`gbrain import` 是从左到右的单向编译，没有反向的"数据库改了同步回 markdown"这回事。

&emsp;&emsp;那"改了源数据库会自动更新吗"这个验收问题，实验室里真跑过一遍，结论很干脆：**不会**。下面这张实验截图把全过程记录下来——改 markdown 前数据库 chunk 的时间戳、用 `echo` 追加一段新内容、改完直接查库（数据库纹丝不动、还是旧时间戳）、重跑 `import` 后再查（这才同步）。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174258058.png" width=80%></div>

&emsp;&emsp;实验分四步，结论一目了然：改 markdown 之前，数据库里那页 chunk 是旧时间戳；用 `echo` 给 markdown 追加一段新内容；**改完直接查库——数据库纹丝不动**，时间戳没变、新内容也没进去，此刻数据库是过期的（stale）；只有重跑 `gbrain import` 之后再查，时间戳才更新、新内容才出现。

&emsp;&emsp;所以验收问题的答案明确了：**改了 git 里的 markdown，数据库不会自动更新，必须重新 `gbrain import`（或增量 `gbrain sync`）才会同步。** GBrain 不监听文件变化——数据库是被动的编译产物，源变了得显式触发"重新编译"。这里有个实用要点：`updated_at` 时间戳是判断库是否过期的快捷方式，拿它和 markdown 的 git 提交时间一比，就知道该不该重 `import`。

### 5.2 PGLite 嵌入式引擎

&emsp;&emsp;数据库这一轨，GBrain 默认用的引擎叫 **PGLite**。这个名字值得你专门记一下，因为它解释了为什么 GBrain 能"零配置零服务器"地在你笔记本上跑起来。

&emsp;&emsp;PGLite 是把 PostgreSQL 编译成 WebAssembly（WASM）后得到的"嵌入式 Postgres"——它不需要你单独启动一个数据库服务进程，而是直接在 Bun 进程内运行，数据落在本地一个目录里。好处很直接：你有完整的 SQL 能力（它就是个真 Postgres），却不用装数据库服务器、不用配端口、不用管账号密码。对个人知识库这种单机场景，这是最省心的选择。

> **【关于 PGLite 底层 Postgres 版本】**：PGLite 官方文档（pglite.dev）只声明它是"WASM Postgres build"，并未在正文标注具体的 Postgres 大版本号。不过我们可以从一个硬证据里看出来——PGLite 建库后，数据目录里有个标准 Postgres 都会有的 `PG_VERSION` 文件，实验室在 v0.42.x 环境实测它的内容是 `17`。所以严谨的说法是："实验室实测 PGLite 数据目录 PG_VERSION 为 17"，而不是"官方文档声明 Postgres 17"——这两者的证据强度不一样，技术写作里要分清。

&emsp;&emsp;说完原理，我们动手建一个带 embedding 的本地 brain。整个过程就**三步**：**① 准备环境 → ② 建库开 embedding → ③ 确认库建在哪**。建好之后，第 5.4 节再往里导数据、算向量。**顺序照着跑即可；中途若卡住，跳到本节末尾的「5.2.1 常见坑排查」对号入座，不要在主线里乱试。**

&emsp;&emsp;**第 ① 步：准备环境（加载凭证 + 定隔离路径）。** 建库这一步本身**不需要 key**，但本章后续 `import` 算向量要用 DashScope、第 7 章 `think` 合成要用 DeepSeek，所以先一次性把两把 key 加载好。库我们统一建在课件目录下的 `./output/brain-l2`——隔离路径，不污染你已有的大脑，产出也方便随时查看和清理。

In [15]:
# 环境准备：本节后续所有 cell 复用此处加载的 key 与变量，不重复加载
from dotenv import load_dotenv
import os

# 从 .env 读取凭证（DashScope 做 embedding，DeepSeek 做 think 合成）；override=True 让 .env 覆盖已有环境变量 （需要安装 python-dotenv 依赖！）
load_dotenv(override=True)

# 第 4 章的 RECAP_PATH 是全程 --no-embedding 的回顾库；从本章起我们另建一个带 embedding 的新库，
# 统一用下面的 BRAIN_PATH，后面所有命令都指向它（两者是不同的库，互不影响）
os.makedirs("output", exist_ok=True)
BRAIN_PATH = "./output/brain-l2"

# 确认 key 已就位（只打印有无，绝不打印 key 值本身）
print("DASHSCOPE_API_KEY:", "已就位" if os.getenv("DASHSCOPE_API_KEY") else "缺失")
print("DEEPSEEK_API_KEY: ", "已就位" if os.getenv("DEEPSEEK_API_KEY") else "缺失")
print("隔离 brain 路径:   ", BRAIN_PATH)

DASHSCOPE_API_KEY: 已就位
DEEPSEEK_API_KEY:  已就位
隔离 brain 路径:    ./output/brain-l2


&emsp;&emsp;这个 cell 把两把 key 和隔离路径一次性备齐：DashScope 给 embedding 用，DeepSeek 留给第 7 章的 `think`。注意只打印"已就位/缺失"而不打印 key 本身——这是处理凭证的基本纪律。**如果打印出"缺失"**：先确认 `.env` 文件的位置（`load_dotenv(override=True)` 会从当前目录向上递归查找 `.env`），key 没加载好，第 ② 步之后的 `import` 一定会失败。

&emsp;&emsp;**第 ② 步：建库（开 embedding）。** `gbrain init --pglite` 在本地初始化一个 PGLite 数据库作为 brain，建库时指定 embedding 模型——我们选 DashScope 的 `text-embedding-v3`（1024 维，对中文优化）。

In [18]:
# 建库：--pglite 用本地 WASM Postgres，--force 允许覆盖重建
# --embedding-model 指定向量模型，--embedding-dimensions 必须显式给（理由见末尾「常见坑排查」）
# 直接用命令行 gbrain（! 前缀，跑在第 3 章装好的全局 gbrain 上）；只看关键行
!gbrain init --pglite --force --path {BRAIN_PATH} --embedding-model dashscope:text-embedding-v3 --embedding-dimensions 1024 --non-interactive 2>&1 | grep -iE 'embedding|schema version|ready|engine'

  Embedding: dashscope:text-embedding-v3 (1024d)
Brain ready at ./output/brain-l2
0 pages. Engine: PGLite (local Postgres).
      the search focuses on what is NEW vs already-known.


&emsp;&emsp;两个参数值得说明：`--pglite` 选本地 WASM Postgres 引擎，数据落在 `--path` 指定的目录；`--embedding-dimensions 1024` **必须显式声明**——不给的话 GBrain 可能从旧 schema 读到一个默认维度（实验室环境残留的是 1536），和 `text-embedding-v3` 实际的 1024 维对不上，`init` 直接报维度冲突而失败。**建库成功的标志**：输出里出现 `Embedding: dashscope:text-embedding-v3 (1024d)`，以及一长串 schema 迁移（`Schema version 1 → 117`，即 GBrain 一次性建好了一百多个表结构）。

&emsp;&emsp;顺带记住第 4 章讲的「单活动库」机制：这条 `init --path {BRAIN_PATH}` 一跑，**活动库就从第 4 章的 RECAP 库切到了这个新的 `BRAIN_PATH` 库**——本章后续所有命令（`import` / `stats` / `query` / `think`）都默认落在它上面，不必再逐条指定库路径。

&emsp;&emsp;**第 ③ 步：确认库建在哪（看它的物理形态）。** 库建好了，落一个最实在的问题——**这个 brain 到底存在硬盘哪个文件夹？** 答案就是你 `--path` 指定的目录（本节是 `./output/brain-l2`；不带 `--path` 用默认库则在 `~/.gbrain/brain.pglite/`）。打开它看一眼：

In [19]:
# 看 PGLite 库的物理落点：--path 指定的目录就是一个完整的 Postgres 数据目录
!ls {BRAIN_PATH} && echo "--- PG_VERSION ---" && cat {BRAIN_PATH}/PG_VERSION

PG_VERSION           pg_multixact         pg_tblspc
base                 pg_notify            pg_twophase
global               pg_replslot          pg_wal
pg_commit_ts         pg_serial            pg_xact
pg_dynshmem          pg_snapshots         postgresql.auto.conf
pg_hba.conf          pg_stat              postgresql.conf
pg_ident.conf        pg_stat_tmp
pg_logical           pg_subtrans
--- PG_VERSION ---
17


&emsp;&emsp;你会看到 `base/`（表数据，子目录是各 database 的 OID）、`global/`、`pg_wal/`、`postgresql.conf`、`PG_VERSION` 这些**真 Postgres 才有的东西**，`PG_VERSION` 内容正是 callout 里说的 `17`。这把"PGLite 就是个编译成 WASM 的真 Postgres"落到眼见为实——它跟本地装的 PostgreSQL 在磁盘上长得一模一样，区别只是跑在 Bun 进程内、由 GBrain 托管。（**Postgres / Supabase 引擎则不同**：数据不在本地文件夹，而在远程服务器上，本地只在 `~/.gbrain/config.json` 的 `database_url` 字段存一个连接串——紧接着 5.3 讲。）

&emsp;&emsp;三步走完，一个带 embedding 的本地库就建好了。**没卡住的话，可以直接跳到 5.3**；下面这一小节是建库阶段三个高频坑的排查手册，卡住了再对号入座。

#### 5.2.1 常见坑排查（建库阶段，对号入座）

&emsp;&emsp;**坑 A · 配了 embedding，库却显示没开（`Embedded: 0` / `doctor` 里 `embed 0/35`）。** 在 v0.42.44.0 上，`gbrain init` 即便带了 `--embedding-model` 也可能把库建成 "deferred"（延迟嵌入）状态——embedding 实际没开。

> **【根因与诊断】**：GBrain 的 config 里有个 `embedding_disabled` 字段，如果之前建过库留下了 `embedding_disabled: true` 残留，这次 `init` 的 `--embedding-model` 会被静默忽略。**诊断**：`cat ~/.gbrain/config.json | grep embedding_disabled`，看它是不是 `true`（config 是扁平 JSON，`embedding_disabled` / `embedding_model` 都在顶层）。注意：事后想用 `gbrain config set embedding_model ...` 补救对 PGLite 是 "silent no-op"（静默空操作），行不通。（v0.42.44 实测复现：把 `embedding_disabled` 设成 `true` 后，`init` 即便带 `--embedding-model` 也只输出 `--no-embedding: deferred setup`、嵌入并未开启。）

&emsp;&emsp;解法看你处在哪种情况。**情况 ①：你刚发现 config 是 `true`、还没建过带数据的库**——跑下面这段把 `embedding_disabled` 改回 `false`（幂等无害，放心跑），再回到第 ② 步重跑 `init`：

In [17]:
# 一键修复 embedding_disabled 残留：改回 false（幂等无害）
import json, os
cfg = os.path.expanduser("~/.gbrain/config.json")
c = json.load(open(cfg)); c["embedding_disabled"] = False
json.dump(c, open(cfg, "w"), ensure_ascii=False, indent=2)
print("embedding_disabled →", c["embedding_disabled"])

embedding_disabled → False


&emsp;&emsp;**坑 B · 已有旧库，改完想重导却全被 skip（`imported=0 skipped=N`）。** 这是上面的**情况 ②**：你已经 `import` 过数据、库里有一批没算向量的页面，光改 config 不够，还得把旧库**清空重建**。**注意别用 `gbrain init --pglite --force`**——GBrain 检测到 embedding 维度要"从无到有"切换是破坏性操作，会拒绝自动清空、只打印 `Switching dims is destructive: it drops every embedding` 警告。正解是官方的 wipe-and-reinit 命令 **`gbrain reinit-pglite`**（专为换 embedding 模型/维度设计，会留 `.bak` 备份可回滚），Notebook 非 TTY 环境**必须加 `--yes`** 跳过确认：

In [ ]:
# 【撞了 import skip 坑才跑】已有 no-embedding 库换成带 embedding：reinit-pglite 一键 wipe+重建
# 破坏性操作，会清空当前活动库（config.database_path，留 .bak 可回滚）；确认要重建再取消注释运行
# 若它提示 ".bak already exists"，先 rm -rf output/brain-l2.bak 删掉旧备份再重跑
# !gbrain reinit-pglite --embedding-model dashscope:text-embedding-v3 --embedding-dimensions 1024 --yes
# 清空后回到 5.4 节重新 import，这次就会 imported=N、不再 skip，且每个 chunk 都真算向量
print("如撞 import skip：取消注释上面的 reinit-pglite（会 wipe 重建带 embedding 的空库），再按 5.4 节 import")

&emsp;&emsp;**坑 C · `import` 算向量报 `invalid_api_key`（用国内/百炼 Key 几乎必撞）。**

> **【根因与解法】**：中国区申请的 `DASHSCOPE_API_KEY` 第一次真连去算向量，很可能报 `invalid_api_key`——**不是 key 错了**，而是 GBrain 内置的 DashScope 配置默认把请求发往国际端点 `dashscope-intl.aliyuncs.com`，中国区 key 在国际端点上会被判无效。**解法**：在 `~/.gbrain/config.json` 加一个 `provider_base_urls` 字段，把 DashScope 基址覆盖成国内端点 `https://dashscope.aliyuncs.com/compatible-mode/v1`——这是 GBrain 提供的正规覆盖机制（可为任意 provider 指定自定义基址，不必改源码）。**排查锚点**：`gbrain doctor` 或下一节 `import` 报 `invalid_api_key` 时，第一个怀疑对象就是端点没改成国内，而不是 key 本身。

### 5.3 Postgres 引擎与"两引擎一契约"

&emsp;&emsp;PGLite 适合个人单机，但如果是团队多人共享、或者数据规模很大，GBrain 还支持连一个真正的 Postgres / Supabase 服务器。这就引出一个很漂亮的设计：**两个引擎，一套契约**。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175016368.png" width=55%></div>

&emsp;&emsp;这套设计的核心是 GBrain 内部有一个叫 `BrainEngine` 的抽象层，它定义了约 90 个操作（put / get / search / query / embed 等等）。PGLite 和 Postgres 各自实现这套接口，于是上层的应用代码、CLI 命令对底层用哪个引擎是无感知的——你今天在 PGLite 上跑通的命令，明天迁到 Postgres 上一字不改照样能跑。这就是"换引擎不换代码"。

&emsp;&emsp;两个引擎的适用场景对照如下：

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>PGLite vs Postgres 引擎选型</font></p>
<div class="center">

| 维度 | PGLite（默认） | Postgres / Supabase |
|------|---------------|---------------------|
| 部署形态 | 嵌入式 WASM，零服务器 | 独立数据库服务器 |
| 就绪速度 | 约 2 秒（进程内启动） | 取决于服务器 |
| 适用场景 | 个人单机、本地笔记 | 团队多人、大规模、生产 |
| 并发写 | 单写者约束（进程独占） | 支持多写并发 |
| 后台 worker | 不支持（见第 8 章 Minions） | 支持 worker daemon |
| 网络攻击面 | 无（不监听端口） | 有（需配认证） |

</div>

&emsp;&emsp;表里有两个细节后面会用到：一是 PGLite 的"单写者约束"——它是进程独占的，不能多个进程同时写；二是"后台 worker 不支持"——这直接关系到第 8 章 Minions 队列的诚实边界（夜间自动维护在 PGLite 上跑不了，得靠别的办法）。我们建库时用的 `PG_VERSION` 文件证据也能在这里复用：PGLite 的数据目录和真 Postgres 长得一模一样，所以两引擎共享同一套 SQL 能力不是巧合，是 PGLite 本就是个完整 Postgres。

#### 5.3.1 动手：把大脑迁到 Postgres，给团队用

&emsp;&emsp;上面是概念。如果你想真的用 Postgres 给团队部署一个共享大脑，下面是一套可以照着做的步骤。注意这部分需要你**自备一个带 pgvector 的 Postgres**（Supabase 托管最省事，也可以自托管），所以下面的命令都用注释保护，你按自己的连接串取消注释运行。

&emsp;&emsp;**第一步：准备 Postgres + pgvector。** GBrain 的向量检索依赖 PostgreSQL 的 **pgvector** 扩展——这是硬依赖，没有它 schema 迁移会直接拒绝运行。三条路：

&emsp;&emsp;① **Supabase（官方主推）**：去官网 [supabase.com](https://supabase.com) 注册一个 Supabase 项目（控制台入口 [supabase.com/dashboard](https://supabase.com/dashboard)），在后台 `Database → Extensions` 里手动开启 `vector` 扩展（顺手把 `pg_trgm`、`pgcrypto` 也开了，分别用于模糊 slug 解析和 UUID）。官方文档特意提醒：这是 5 秒的事，忘了开会让你 debug 一小时。

&emsp;&emsp;② **自托管 Postgres**：确保你的 Postgres 装了 pgvector，且连接用的数据库角色有建扩展权限。

&emsp;&emsp;③ **本地 Docker（零成本，最适合本机快速验证）**：不想注册 Supabase、也不想在本机直接装 Postgres，最干净的办法是用官方 `pgvector/pgvector` 镜像起一个**自带 pgvector** 的 Postgres——一条命令搞定、数据持久化到本地目录、用完即弃。**注意不能用纯 `postgres` 官方镜像**：它不含 pgvector，`init` 建 `vector` 扩展会直接失败，必须用 `pgvector/pgvector`。

In [ ]:
# 本地 Docker 起 pgvector（需先装 Docker Desktop 并启动）；按需取消注释逐条运行
# ① 起一个带 pgvector 的 Postgres，数据持久化到本地 ~/Docker 目录（用完 docker rm -f gbrain-pg 即弃）
# !docker run -d --name gbrain-pg -e POSTGRES_PASSWORD=postgres -p 5432:5432 -v ~/Docker/gbrain-pg-data:/var/lib/postgresql/data pgvector/pgvector:pg17

# ② gbrain 连接它（schema 自动迁移 1→117 + 自动 CREATE EXTENSION vector，活动库切到 postgres）
# !gbrain init --url "postgresql://postgres:postgres@localhost:5432/postgres" --non-interactive

# ③ 之后所有命令自动作用在这个 postgres brain（单活动库机制，不必带库参数）
# !gbrain import {WIKI_DIR} --no-embed
# !gbrain stats
print("本地 Docker 方案已注释保护；装好 Docker Desktop 后取消注释逐条运行")

&emsp;&emsp;容器跑起来后，在 Docker Desktop 的 Containers 面板里就能看到这个 `gbrain-pg` 容器——镜像 `pgvector/pgvector`、端口 `5432:5432`、状态绿点运行中（下图红框）。这一套是实测可跑的：`gbrain init --url` 会打印 `Schema version 1 → 117` 迁移成功、`Engine: Postgres`，之后 `import` / `stats` 和 PGLite 一字不差地照常工作——正是 5.3 开头说的"换引擎不换代码"。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174229714.png" width=70%></div>

&emsp;&emsp;`gbrain init` 连库时会**自动尝试** `CREATE EXTENSION IF NOT EXISTS vector`；如果当前角色没权限，它会降级提示你手动去 SQL Editor 跑 `CREATE EXTENSION vector;`（这是源码 `init.ts` 里实测确认的行为，不是猜的）。

&emsp;&emsp;**第二步：连接 Postgres。** 有三种写法，按场景选一个：

In [20]:
# 注意：以下需自备 Postgres/Supabase，用注释保护；按你的实际连接串取消注释运行

# 写法一：自托管 Postgres，命令行直接给连接串（最直接）
!gbrain init --url "postgresql://postgres:postgres@localhost:5432/postgres"

# 写法二：Supabase，走交互向导（它会一步步引导你贴 connection string）
# !gbrain init --supabase

# 写法三：CI / 脚本部署，用环境变量 + 非交互模式
# import os; os.environ["GBRAIN_DATABASE_URL"] = "postgresql://..."
# !gbrain init --non-interactive

print("连接配置已注释保护；自备 Postgres 后，按写法一/二/三取消注释运行")

  Embedding: dashscope:text-embedding-v3 (1024d)
  Expansion: openai:gpt-5.2
Connecting to database...

Refusing to silently re-template existing brain.

  Existing column: vector(1280)
  Requested:       vector(1024)  (dashscope:text-embedding-v3)

Switching dims is destructive: it drops every embedding in your brain and
requires a full re-embed (potentially hours and $1-100 in API calls).

Recipe (run against your Postgres brain):

  BEGIN;
  DROP INDEX IF EXISTS idx_chunks_embedding;
  -- NULL embeddings BEFORE the alter: pgvector refuses to cast existing
  -- vectors across dimensions and aborts the transaction. NULLs cast fine.
  UPDATE content_chunks SET embedding = NULL, embedded_at = NULL;
  ALTER TABLE content_chunks ALTER COLUMN embedding TYPE vector(1024);
  CREATE INDEX IF NOT EXISTS idx_chunks_embedding
    ON content_chunks USING hnsw (embedding vector_cosine_ops);
  COMMIT;

Then re-init config (file plane is canonical post-v0.37):
  gbrain init --supabase --embedding-mo

&emsp;&emsp;还有个 PGLite 给不了的福利：因为 Docker Postgres 是**真服务器、监听 5432 端口**，你可以直接用 `psql` 连进去查（对比 4.4 节 PGLite 是进程内 WASM、不监听端口，只能用 JS 接口）——一条 `docker exec gbrain-pg psql -U postgres -c "SELECT slug,type,title FROM pages"` 就能列出库里已导入的页面。下面真跑几条只读 SQL 看看（前提：上面 ① docker 起容器、② `gbrain init`、③ `gbrain import` 都已执行，库里有数据）：

In [ ]:
# 用真 SQL 直连 Docker Postgres 查表（前提：上面 5.3 节起的 gbrain-pg 容器在跑）
# Docker Postgres 监听 5432、是真服务器 → 能 docker exec 进容器跑 psql；PGLite 不监听端口，做不到

# ① 列出库里所有页面：slug / 类型 / 标题（就是上面正文点名的那条命令）
!docker exec gbrain-pg psql -U postgres -c "SELECT slug, type, title FROM pages ORDER BY type, slug"

# ② 用 GROUP BY 按类型统计页数（concept / debate / note / system 各几页）
!docker exec gbrain-pg psql -U postgres -c "SELECT type, count(*) FROM pages GROUP BY type ORDER BY count DESC"

# ③ 看 chunk 与向量化情况：本节 import 用了 --no-embed，所以 embedded 为 0
!docker exec gbrain-pg psql -U postgres -c "SELECT count(*) AS chunks, count(embedding) AS embedded FROM content_chunks"

       slug        |  type   |           title            
-------------------+---------+----------------------------
 compiled-rag      | concept | 编译式 RAG
 compiled-rag-cost | debate  | 编译式 RAG 的成本争议
 index             | note    | Wiki 内容目录
 log               | note    | Ingest 日志（append-only）
 gbrain            | system  | GBrain
(5 rows)

  type   | count 
---------+-------
 note    |     2
 debate  |     1
 concept |     1
 system  |     1
(4 rows)

 chunks | embedded 
--------+----------
      8 |        0
(1 row)


&emsp;&emsp;三条 SQL 的输出印证了几件事：① `pages` 表里 5 页（concept / debate / note×2 / system）原样躺着，`SELECT` 和查任何 Postgres 表没有区别；② `GROUP BY` 这种标准 SQL 聚合也照常能用——因为 GBrain 的库就是一个普通 Postgres，没有任何私有查询协议；③ `content_chunks` 有 8 个 chunk 但 `embedded` 是 0，正是因为上面 `import` 带了 `--no-embed`（向量留到 5.4 节再补）。这正是 Postgres 引擎相比 PGLite 多出来的那条路：**任何 SQL 客户端都能直连进来观察、审计、二次分析**，而 PGLite 不监听端口、只能隔着 gbrain CLI 看。

&emsp;&emsp;连接串会写进 `~/.gbrain/config.json` 的 `database_url` 字段，引擎记为 `postgres`。这里有个 **Supabase 专属的坑**要先讲清：Supabase 的连接串要用 **Transaction pooler（端口 6543）**，而不是 direct 连接（端口 5432，它是 IPv6-only 的，很多机器连不上）；如果有需要直连的操作（数据库迁移、worker 锁），再额外配一个 `GBRAIN_DIRECT_DATABASE_URL` 指向 5432。**自托管的 Postgres 没有这个 IPv6 问题**，一个 `--url` 就够了。

&emsp;&emsp;**第三步（推荐路径）：先 PGLite 起步，长大了一条命令迁过去。** 你不必一开始就上 Postgres。最省事的做法是：先用 PGLite 单机把大脑建好、跑顺，等数据量大了或者要多人共享，再迁移到 Postgres：

In [ ]:
# 把当前 PGLite 大脑迁到 Supabase/Postgres（搬运页面、向量、图谱边、tags、timeline）
# !gbrain migrate --to supabase --url "postgresql://..."

# 反向迁回 PGLite（少见，比如想把团队库拉到本地调试）：
# !gbrain migrate --to pglite --path ./output/brain-local

print("迁移命令已注释保护；gbrain migrate --to <supabase|pglite> 在两个引擎间搬运大脑的主要数据")

&emsp;&emsp;`gbrain migrate --to supabase` 把 PGLite 里的页面、向量、图谱边整体搬到 Postgres（命令在源码 `migrate-engine.ts` 里核对过真实存在）。迁完之后，你前面学的所有命令——`import` / `extract links` / `query` / `think`——一字不改照样跑，这就是 5.3 节开头说的"换引擎不换代码"。

&emsp;&emsp;**第四步：团队共享的两个硬约束。**

&emsp;&emsp;① **后台 worker 必须 Postgres。** 第 8 章会讲的"夜间自动维护"（`gbrain jobs work` worker 守护进程）在 PGLite 上根本起不来——PGLite 的独占文件锁挡住了独立 worker 进程。所以**任何需要后台队列、自动维护的团队场景，Postgres 是前置条件**，这是硬约束不是建议。

&emsp;&emsp;② **多人共享走 MCP + 认证，不是把连接串发给每个人。** GBrain 团队大脑的正规姿势是起一个 HTTP MCP 服务（`gbrain serve --http`），每人用自己的客户端接入，数据隔离在 SQL 层按 source 强制划分读写范围——而不是把数据库连接串群发出去。这部分属于团队运维进阶，本节点到为止，知道"正规做法是 MCP 服务 + 按 source 隔离"即可。

> **【诚实边界】**：本节的 Postgres 命令都经过源码逐条核对、确认真实存在，但因为需要一台真实的 Postgres 服务器，我们**没有在课件环境里实跑**（前面 PGLite 那部分是全程真跑的）。另外两点官方也没写死，提醒你别照着死磕：**Postgres 最低大版本号**官方文档未明确规定（源码间接暗示 13+ 可用），**pgvector 的版本下限**也没指定——实践上用现代 Postgres（13/14 以上）配最新 pgvector 即可。

### 5.4 开启 embedding 向量化

&emsp;&emsp;到这里，我们终于要做前面刻意跳过的那件事了——**开 embedding**。前面全程 `--no-embedding`，知识进了库但没有语义向量，所以只能做关键词级别的事。这一节我们把向量补上，知识才真正"带上语义坐标"，第 6 章的向量检索才有了燃料。

&emsp;&emsp;embedding 做的事，一句话说清：把每段文本转成一组数字坐标（我们这里是 1024 个浮点数），语义相近的文本，坐标方向也相近。这样查询时，把问题也编码成同样的坐标，在库里找"方向最接近"的若干段，就是向量检索。这跟图书馆给每本书贴多维度标签、再在标签空间里找最近邻是同一个直觉。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174425943.png" width=55%></div>

&emsp;&emsp;我们的隔离 brain 刚建好还是空的，先导入一批语料。这里用一份现成的 wikilink 数据（3 个人物 + 3 家公司，彼此用 wikilink 连着），它正好适合后面演示"跨页综合"。

In [21]:
# 导入语料目录：6 页 markdown（people/ 3 人 + companies/ 3 公司，带 wikilink 关系）
CORPUS_DIR = "./gbrain-wikilink-data/"

# 扫描目录所有 markdown，切 chunk + 算 1024 维向量入库；只看导入汇总
!gbrain import {CORPUS_DIR} 2>&1 | grep -iE 'imported|chunks created|Import complete'

[import.files] 4/6 (66%) imported=4 skipped=0 errors=0
[import.files] 6/6 (100%) imported=6 skipped=0 errors=0
Import complete (1.4s):
  6 pages imported
  6 chunks created


&emsp;&emsp;`gbrain import` 扫描目录下所有 markdown，为每页切 chunk、为每个 chunk 调 embedding API 算 1024 维向量，写进数据库。因为开了 embedding，这一步会真实发起网络请求算向量，所以比前面 `--no-embedding` 时慢一点。导入完你会看到 `6 pages imported, 6 chunks created`——6 页都成功，每页一个 chunk。

&emsp;&emsp;导入后立刻验证一下入库情况，这是一个 Tier 1 级别的结构验证。

In [22]:
# Tier 1 验证：确认 6 页都进了库，列出 slug / 类型 / 标题
!gbrain list -n 10

companies/alphaventures	company	2026-06-24	AlphaVentures
companies/nexaflow	company	2026-06-24	NexaFlow
companies/sparkfield	company	2026-06-24	SparkField
people/alice-chen	person	2026-06-24	Alice Chen
people/bob-morgan	person	2026-06-24	Bob Morgan
people/carol-wu	person	2026-06-24	Carol Wu


&emsp;&emsp;输出会列出 6 行：3 家公司（alphaventures / nexaflow / sparkfield）和 3 个人物（alice-chen / bob-morgan / carol-wu），每行带类型和标题。这证明语料连同它们的 embedding 都已经入库，第 6 章的向量检索有底料了。

&emsp;&emsp;再确认一下 embedding 真的算好了——用 `gbrain doctor` 看健康分。

In [23]:
# 用 doctor 看 brain 健康分，重点看 embed 子项是否满分（只抓 brain_score 行）
!gbrain doctor 2>&1 | grep -iE 'brain[_ ]score'

  [WARN] brain_score → Brain score 45/100 (embed 35/35, links 0/25, timeline 0/15, orphans 0/15, dead-links 10/10)
  [WARN] brain_score: Brain score 45/100 (embed 35/35, links 0/25, timeline 0/15, orphans 0/15, dead-links 10/10)
Weighted brain score: [WARN] Brain score 45/100 (embed 35/35, links 0/25, timeline 0/15, orphans 0/15, dead-links 10/10)


&emsp;&emsp;你会看到类似 `Brain score 45/100 (embed 35/35, links 0/25, timeline 0/15, orphans 0/15, dead-links 10/10)` 的输出。重点看 `embed 35/35`——embedding 子项满分，说明 6 页的向量都算好了。至于总分只有 45，`links 0/25` 这一项要解释清楚：**它不是"边建不出来"，而是本节根本没去抽边**——图谱的 typed edges 要单独跑一次 `gbrain extract links` 才会建（self-wiring 自动建图是另一条独立的线），本节我们专注"存储 + 检索"这条线，没碰建图那一步，所以 links 自然是 0；`timeline 0/15` 同理（语料没有时间线信息）。这些都不影响向量检索——这一步我们要的就是 `embed 35/35` 这个满分。

&emsp;&emsp;最后认识一个最常用的"库信息总览"命令——`gbrain stats`，一行看清库里的全部家底：

In [24]:
# gbrain stats：库的整体统计（页数 / chunk 数 / 已嵌入数 / 图谱边 / 标签 / 时间线 + 按类型分布）
!gbrain stats

Pages:     6
Chunks:    6
Embedded:  6
Links:     0
Tags:      0
Timeline:  0

By type:
  person: 3
  company: 3


&emsp;&emsp;输出会显示 `Pages` 和 `Chunks` 都是 6、`Embedded` 也是 6（6 页向量都算好了，与上面 `embed 35/35` 互相印证），`Links` 是 0（本节没抽边）。这里补一个**访问方式上的硬约束**：PGLite 虽然是完整 Postgres，但它跑在进程内、不监听端口（数据目录 `postgresql.conf` 里 `port` 和 `listen_addresses` 都是注释掉的），所以你**没法用 `psql` 从外部直连**——库里的信息只能隔着 GBrain 用 `stats` / `list` / `doctor` 这类命令看。这正是 PGLite "零服务器、无网络攻击面"（5.3 表里那一行）的另一面。

&emsp;&emsp;Storage 层到这里补全了：你理解了"源真相 vs 编译产物"的双轨关系、PGLite 和 Postgres 两引擎一契约的设计、第一次把 embedding 开了起来。知识带上语义向量之后，下一站我们就能深入前面完全没碰的 Retrieval 层，看 GBrain 是怎么把对的页面捞出来排好序的。

---

## <center>第 6 章：Retrieval 检索层</center>

&emsp;&emsp;第三站，也是前面完全没碰的一层。此前你的 brain 全程没有 embedding，连向量检索的门槛都没跨过；现在向量补上了，我们终于能问一个核心问题：当你提一个问题，GBrain 到底怎么从库里把相关页面捞出来排好序？答案不是"单跑一路检索"，而是同时跑向量和关键词两路、再用一个叫 RRF 的方法把两路排名融合。这一章把这三路拆开看清楚——每路各补什么短板、融合之后排序怎么变，最后还有两个真实的工具坑实战。6.4 节的 RRF 是本章的难点高峰，我们会先用错的融合方式、再揭示正确的。

### 6.1 两种检索的盲区

&emsp;&emsp;先回答"为什么不单跑一路"。检索一个问题，最直觉的做法是关键词匹配——查询里的词在哪页出现得多，哪页排前面。这是经典 BM25 思路，精确、可解释，但有个硬伤：它只认字面。如果你用"可复利的资产"去问，而页面标题写的是英文 `compounding asset`，纯关键词就漏了。

&emsp;&emsp;另一条路是向量检索：把查询和每页都编码成语义向量、比较接近程度。它能跨同义词、跨语言抓到语义相关页面，但代价是对精确的专有名词、罕见技术词反而不够敏感，还可能把"沾点边但其实无关"的页面也召回。两路各有盲区，用一张表摆清楚。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174902544.png" width=55%></div>

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>向量检索 vs 关键词检索的互补性</font></p>
<div class="center">

| 维度 | 向量检索（HNSW） | 关键词检索（BM25） |
|------|-----------------|-------------------|
| 匹配依据 | 语义相似（向量方向） | 字面词频（精确匹配） |
| 擅长 | 跨同义词、跨语言、抓语义相关 | 专有名词、罕见技术词命中又准又干脆 |
| 盲区 | 对精确词不敏感、可能召回无关页 | 漏同义、漏跨语言表述 |
| 典型分数量纲 | 约 0.7 ~ 1.1（RRF 融合后） | 明显高于 RRF（tsvector 原始分，绝对值随语料规模变化） |
| GBrain 实现 | pgvector + HNSW 近邻 | PostgreSQL tsvector 全文检索 |

</div>

&emsp;&emsp;最后一行那个"分数量纲"的差异先记住——向量这路融合后约 1.0，BM25 那路的 tsvector 原始分明显更高（在 case-4 那种较大语料上最高分能到约 2.9，差快三倍；本节 6 页小语料上量级小一些，但同样高于 RRF 区间）。关键不是记死哪个具体数，而是记住**两路量纲不在一个尺度**——这个差异是 6.4 节 RRF 必须存在的核心理由。GBrain 的回答很干脆：两路都跑，再融合。下面逐路看。

### 6.2 HNSW 向量检索

&emsp;&emsp;向量这一路，GBrain 用的索引算法叫 **HNSW**（Hierarchical Navigable Small World，分层可导航小世界图）。

> **【30 秒回顾向量概念】**：前面第 5 章你接触过 embedding，这里快速接住——把每页文本转成一个 1024 维向量（一组浮点数），语义相近的文本向量方向也相近。检索时把查询也编码成同维向量，找"方向最接近"的若干页。问题是：库里页面多时，逐个比对太慢。HNSW 就是解决"怎么快速找到最近邻"的算法。

&emsp;&emsp;HNSW 的做法是把所有页面的向量组织成一张可快速跳转的分层图，查询向量进来后能在大量向量里快速跳到最接近的若干个，而不必逐一比对——把检索复杂度从线性降到对数级。你不需要会实现它，只要知道 GBrain 的向量检索走的是 HNSW、它让向量检索在大库上也跑得快就够了。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174813452.png" width=55%></div>

&emsp;&emsp;这里要澄清一个最容易卡住的点：**HNSW 到底有没有改变"算相似度"这件事？答案是没有——它底层还是向量相似度计算，一次都没省。** 把"找最近邻"这个动作拆开，对比两种做法就清楚了：

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>暴力法 vs HNSW：相似度计算没变，省的是次数</font></p>
<div class="center">

| 维度 | 暴力法（逐个比对，flat） | HNSW |
|------|------------------------|------|
| 怎么算"相似" | 向量距离（余弦 / 欧氏） | **完全一样，没变** |
| 和多少个向量算距离 | 全部 N 个，一个不漏 | 只和分层图上走过的少数节点（约 log N 个） |
| 速度 | 慢，O(N) 线性 | 快，约 O(log N) |
| 结果 | 精确（真·最近邻） | 近似（可能漏极少数） |

</div>

&emsp;&emsp;关键就在这张表的前两行：**两种做法用的是同一个距离函数、同一种相似度计算，HNSW 完全没改"单次距离怎么算"；它改的只是"要算多少次"**——从"和每一个向量都算一遍"，变成"只和那张分层图上走过的少数几个节点算"。所以更准确的说法是：HNSW 不是"替代了相似度计算的新算法"，而是"**一套让你少算距离的索引结构**"。回到"是不是只加了个最近邻算法"这个疑问——本质对，但精确说：加的是一套**让相似度计算少算几个数量级次数**的结构，单次计算本身一点没动。

&emsp;&emsp;天下没有白省的：HNSW 顺着图**贪心**地走，有极小概率走偏、漏掉那个真正最近的向量，所以它找到的是**近似**最近邻、而非保证精确——这正是下面要说的"ANN"里"近似"二字的由来。这是**用一点点精度换巨大的速度**：暴力法精确但慢，HNSW 可能命中 99.x% 但快几个数量级，库越大这笔交易越划算（建图的 `M`、查询的 `ef_search` 这些参数就是调"算多少、多准"的旋钮：调大更准更慢，调小更快但偶尔漏）。

&emsp;&emsp;再把 HNSW 放回 RAG 的全局里定个位，你就知道它到底是哪个零件。RAG（检索增强生成）拆开就是**检索 + 生成**两步：检索这步通常是向量检索——文档切块算 embedding 存库，查询也算成 embedding，**在库里找最相似的若干块**喂给 LLM。而"在库里找最相似的若干块"这一步，本质就是 ANN（Approximate Nearest Neighbor，近似最近邻）搜索，**HNSW 正是实现它的主流索引**（pgvector、Faiss、Milvus、Qdrant 都用它）。所以记住这个分工：**embedding 决定"哪些向量算相近"（语义质量），HNSW 决定"怎么快速找到这些相近的向量"（检索速度）**——HNSW 不改变检索的质量上限，只让向量检索在大规模库上还跑得动，是 RAG 能落到百万级文档上的关键基础设施。在 GBrain 里，它只是混合检索三件套（HNSW 向量 + BM25 + RRF）里"向量这一路"的加速索引，不是检索的全部。

&emsp;&emsp;我们直接跑一个向量检索看效果。这里有个**容易记反、但必须记住**的细节：`gbrain query` 和 `gbrain search` **默认都是混合检索**（HNSW 向量 + BM25 + RRF，都带语义向量）——它们的区别不是"带不带向量"，而是控制粒度：`query` 是完整可控版，能开关 LLM 查询扩展（`--no-expand` 就是关掉它）；`search` 是轻量版，默认不做查询扩展、成本更低。**只有一种情况它们会退成纯关键词：环境里没有可用的 embedding provider 时**（比如你把 embedding key 剔掉），算不了向量，就只剩关键词这一路——这一点下一节（6.3）演示纯 BM25 时会用到。我们先跑 `query --no-expand`（`--no-expand` 跳过那一步 LLM 查询扩写、防止 #1368 那种 CPU 挂死，6.7 节详讲，这里先记住要加它）。

In [25]:
# 向量/混合检索：gbrain query（HNSW 向量 + BM25 + RRF）
# --no-expand 跳过查询扩展（见 6.7 节 Issue #1368），生产查询建议默认带上
# 跑一个英文 query，看向量怎么跨语言命中中文/英文混合语料
!gbrain query "NexaFlow founder" --no-expand --limit 5

[0.9416] companies/nexaflow -- # NexaFlow
NexaFlow is a workflow startup founded in 2021 by [[people/alice-chen]].
The engineering 
[0.8937] people/alice-chen -- # Alice Chen
Alice Chen is the founder of [[companies/nexaflow]], which she co-founded in 2021.
Alic
[0.8687] people/carol-wu -- # Carol Wu
Carol Wu works at [[companies/nexaflow]] as CTO since 2022.
She manages engineering at Ne
[0.8216] companies/alphaventures -- # AlphaVentures
[[people/bob-morgan]] is a partner at AlphaVentures, leading investments.
AlphaVentu
[0.8094] people/bob-morgan -- # Bob Morgan
Bob Morgan works at [[companies/alphaventures]] as a general partner.
Bob Morgan invest


&emsp;&emsp;输出会是几行 `[分数] slug` 的列表，比如 `[0.9416] companies/nexaflow`、`[0.8937] people/alice-chen`、`[0.8687] people/carol-wu`——向量检索把和"NexaFlow founder"语义相关的页面都召回了，包括公司页和它的创始人页、CTO 页。分数在 0.87~0.94 之间，是融合后的量纲（约 1.0）。这是向量这路的本事：抓语义关联，不死磕字面。

### 6.3 BM25 关键词检索

&emsp;&emsp;关键词这一路用的是 **BM25**（Best Matching 25），一个经典的相关性打分公式，按词频、逆文档频率给"查询词与文档的字面匹配程度"打分。GBrain 用 PostgreSQL 的 `tsvector` 全文检索实现这一路，特点是对精确词、专有名词命中得又准又干脆。

&emsp;&emsp;BM25 对应的命令是 `gbrain search`。但这里有个**必须先讲清的隔离坑**，否则你跑出来的 BM25 数据是假的：

> **【踩坑预警】**：想单独看 BM25 这一路（纯关键词）的真实分数，得先把 embedding key 剔掉——因为**只要环境里有可用的 `DASHSCOPE_API_KEY`，`gbrain search` 默认就走混合检索（向量 + 关键词 + RRF），不是纯 BM25**。判断的关键不是某个绝对分值，而是**量纲风格**：纯 BM25 走 tsvector，分数量纲明显偏高（case-4 大语料上能到约 2.9，本节小语料量级小一些）；混合检索的分数会落到约 1.0 这个 RRF 区间。后果是你以为在测 BM25 一路、其实测的是混合，三路对比直接失真。正确做法是跑纯 BM25 时用 `env -u DASHSCOPE_API_KEY` 把这个变量从子进程环境里剔除——没了 embedding provider，search 就只能退回纯关键词。排查方法：如果你的"BM25 分数"看着落在约 1.0 的混合区间、而不是 tsvector 那种更高的量纲，十有八九是 key 没剔掉、它还在走混合。

&emsp;&emsp;我们把这个隔离落成代码——构造一个剔除了 key 的环境专门跑 BM25。

In [26]:
# 纯 BM25 检索：用 env -u 剔除 DASHSCOPE_API_KEY，search 才会退成纯关键词
# （有 key 时 search 默认走混合检索；剔掉 key 没了 embedding provider，就只剩 BM25 一路）
# 同一个 query，纯 BM25 这路能找到什么（可能返回 No results）
!env -u DASHSCOPE_API_KEY gbrain search "NexaFlow founder" --limit 5

No results.


&emsp;&emsp;这里有个细节要注意：这个 query 跑纯 BM25 可能返回"No results"（无结果）。为什么？因为我们的语料以英文人物/公司页为主，但 `tsvector` 对这个特定 query 的分词与匹配没能命中精确词。对照刚才向量这路（6.2）召回了 3 页——这就直观看到了两路的差异：**向量补上了 BM25 漏掉的语义相关页面**。在 case-4 实验室的更大语料上，这个对比更明显：一个语义型查询，BM25 只命中 2 页，混合检索命中 10 页，向量足足补了 8 页。

### 6.4 RRF 排名融合

&emsp;&emsp;两路结果都有了，现在到本章最难也最关键的一步：怎么把它们融合成一个排名？这一节我们用"先想错、再纠正"的方式来讲，因为最直觉的做法恰恰是错的。

&emsp;&emsp;**第一步：需求。** 我们要把向量一路和 BM25 一路的结果合并成一个统一排序，让用户看到一份榜单。

&emsp;&emsp;**第二步：最直觉的方案——两路分数直接相加。** 每页在向量这路有个分、在 BM25 那路有个分，加起来排序不就行了？听起来天经地义。

&emsp;&emsp;**第三步：暴露缺陷——量纲根本不可比。** 回看 6.1 那张表：BM25 的 tsvector 原始分明显高于向量融合后的约 1.0。这在 case-4 的较大语料实验里看得最清楚——跑同一个精确词查询，BM25 最高分 2.9167、混合最高分 1.0499，这俩根本不在一个尺度上（约差三倍）。本节 6 页小语料上绝对值更小，但量纲偏高的方向一致。重点是：直接相加，等于让 BM25 的分数"权重"天然比向量大一大截——融合结果会被 BM25 单方面主导，向量的意见几乎被淹没。所以**直接加分是错的**。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175016430.png" width=55%></div>

&emsp;&emsp;**第四步：正确方案——RRF 按名次融合。** RRF（Reciprocal Rank Fusion，倒数排名融合）的精妙之处是：**完全不碰各路的原始分数，只看名次**。一个文档在某路排第 `rank` 位，就贡献 `1/(k + rank)` 的分数，把各路贡献加起来作为融合分。公式是：

```
RRF_score(doc) = Σ  1 / (k + rank_i)     对每一路检索 i 求和
```

&emsp;&emsp;`rank_i` 是该文档在第 `i` 路里的名次，`k` 是平滑常数，GBrain 默认 `k = 60`。因为只看名次不看分数，"BM25 分和向量分量纲不可比"这个难题被直接绕开了——不管 BM25 是 2.9 还是 0.4、向量是 1.0 还是 0.3，名次永远是 1、2、3 这种纯序数，天然可比。

&emsp;&emsp;`k = 60` 的作用是抑制低名次项之间的差异：一个文档两路都排第 1，得 `1/61 + 1/61 ≈ 0.0328`；一路第 1、另一路第 10，得 `1/61 + 1/70 ≈ 0.0307`——分数只略低。意思是即便某文档在某一路表现一般，只要另一路认可，仍能拿到接近的融合分。我们用代码把这个"平权"算一遍。

In [27]:
def rrf(ranks, k=60):
    """RRF 融合打分：对一个文档在各路的名次列表求 Σ 1/(k+rank)

    Args:
        ranks: 该文档在各路检索里的名次列表，如 [1, 4] 表示一路第1、另一路第4
        k: 平滑常数，GBrain 默认 60
    Returns:
        融合后的 RRF 分数（float），越大越靠前
    """
    return sum(1.0 / (k + r) for r in ranks)

# 构造两个名次互补的文档，验证 RRF 是否给两路平权
# 对照两个文档：A 在 BM25 第1、向量第4；B 在 BM25 第4、向量第1
doc_a = rrf([1, 4])   # 精确匹配强、语义中等
doc_b = rrf([4, 1])   # 语义中心、精确匹配弱
print(f"Doc A (BM25第1, 向量第4): RRF = {doc_a:.5f}")
print(f"Doc B (BM25第4, 向量第1): RRF = {doc_b:.5f}")
print(f"两者几乎相等 → RRF 给了向量和 BM25 平等投票权")

Doc A (BM25第1, 向量第4): RRF = 0.03202
Doc B (BM25第4, 向量第1): RRF = 0.03202
两者几乎相等 → RRF 给了向量和 BM25 平等投票权


&emsp;&emsp;运行结果会显示 Doc A 和 Doc B 的 RRF 分几乎相等（都约 0.032）。这说明：BM25 的"第 1 名"并没有把 Doc A 锁死在榜首——向量把 Doc B 顶上来，两者打平。这正是融合的意义：不让任何一路单方面说了算。

&emsp;&emsp;这个"重排"在真实数据上看得更清楚。下面是 case-4 实验室的三路打分 DataFrame 截图——一个精确词查询，BM25 把"标题精确匹配"的页排第 1，但 RRF 综合语义后把"语义中心"的页提到了第 1，名次被重排了。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174240631.png" width=70%></div>

&emsp;&emsp;这张表把"向量补语义"摊开了：只有头两行有 BM25 分，其余 8 行 BM25 完全没召回（空值），它们的 RRF 分纯粹来自向量贡献——这就是向量这路最纯粹的证据。下面这张柱状图把格局画出来，红=BM25、蓝=向量、绿=RRF 融合。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174252806.png" width=70%></div>

&emsp;&emsp;左图（语义型查询）大部分页面只有蓝绿柱、红柱稀疏——这是"向量补语义"；右图（精确词查询）红柱高但绿柱顺序不同——这是"RRF 重排"。到这里，"BM25 漏语义 → 向量补充 → RRF 融合"的完整原理就可视化地立住了。

&emsp;&emsp;我们也亲手跑一个中文 query 验证向量的跨语言能力——语料是英文页，但我们用中文问。

In [28]:
# 中文 query 检索英文语料，验证向量的跨语言语义能力（Tier 2 端到端）
!gbrain query "谁投资了这些创业公司" --no-expand --limit 3

[0.8630] companies/alphaventures -- # AlphaVentures
[[people/bob-morgan]] is a partner at AlphaVentures, leading investments.
AlphaVentu
[0.8434] companies/sparkfield -- # SparkField
[[people/bob-morgan]] invested in SparkField as part of [[companies/alphaventures]]'s p
[0.8310] people/bob-morgan -- # Bob Morgan
Bob Morgan works at [[companies/alphaventures]] as a general partner.
Bob Morgan invest


&emsp;&emsp;输出会命中 `companies/alphaventures`、`companies/sparkfield`、`people/bob-morgan` 这几页（分数约 0.83~0.86）——我们用中文问"谁投资了"，向量跨语言找到了英文的投资方页面。这是纯 BM25 绝对做不到的（中文 token 在英文页里没有字面匹配），是混合检索的硬价值。

### 6.5 RRF 之上的四层加权

&emsp;&emsp;RRF 给了基础排名，但 GBrain 在它之上还叠了四层加权，让排序更贴合真实需求。这些在演示小语料上看不太出差异，但理解它们对调优很重要，用一张速查表过一遍。

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>RRF 之上的四层加权</font></p>
<div class="center">

| 加权层 | 做什么 | 典型场景 |
|--------|--------|----------|
| source-tier boost | 按来源可信度加权：手写 wiki（Tier 1）抬升，自动导入（Tier 2）、草稿（Tier 3）降权 | 多源 brain，区分可信度 |
| 图邻接加权 | 和命中页有 wikilink 关系的邻接页适当加分 | 利用 self-wiring 建的图谱 |
| 意图感知查询改写 | 理解查询意图后改写检索词 | 模糊查询提质 |
| ZeroEntropy 重排 | 末段精排，进一步优化 top 结果顺序 | 追求高精度排序 |

</div>

&emsp;&emsp;这里有个诚实边界要说：**source-tier boost 在单源 brain 上看不出差异**。我们的演示语料全是同一来源（手写 wiki），tier 全相同，这层加权没有可见效果。要看到它，需要多源场景——比如分别 import 一批可信文档和一批草稿笔记，再观察排序差异。我们这节课不展开多源，知道有这层、知道它在单源场景隐身就够了。

### 6.6 三档成本质量模式

&emsp;&emsp;检索还有一个调节旋钮：成本/质量档位。GBrain 提供三档，让你在"省钱"和"质量"之间选。

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>检索三档模式对照</font></p>
<div class="center">

| 档位 | 取向 | 适用 |
|------|------|------|
| conservative | 省成本，少调用 | 成本敏感、可接受质量略降 |
| balanced（默认） | 成本与质量平衡 | 大多数场景 |
| tokenmax | 质量优先，多读多算 | 追求最佳综合质量、不在乎成本 |

</div>

&emsp;&emsp;默认是 balanced，对大多数场景够用。这个档位的取舍逻辑会在第 9 章 benchmark 那里再次呼应——"成本贵不等于质量好"，档位选择本质是在成本和质量之间做权衡。

### 6.7 两个工具坑：#1368 与 key 隔离

&emsp;&emsp;本章最后是两个真实的工具坑实战。前面所有 `gbrain query` 命令我们都带了 `--no-expand`，现在来讲清它是干什么的，以及它对应的那个著名 Issue。

&emsp;&emsp;**坑一：Issue #1368——默认查询可能 CPU 挂死。** `gbrain query` 有一个默认开启的多查询扩展（multi-query expansion）功能：在真正检索之前，先调一个语言模型把你的原始查询"扩写"成若干变体，再分别检索、最后融合。出发点是好的（多角度提高召回），但它引入了一次额外的 LLM 调用——当这个扩展模型不可用或响应很慢时，`gbrain query` 可能 CPU 飙到接近 100%、迟迟不返回，且没有内置超时兜底。这就是 Issue #1368。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175042387.png" width=70%></div>

&emsp;&emsp;解法就是 `--no-expand`：跳过扩展步骤，查询编码后直接走 HNSW + BM25 + RRF，立即返回。它不是临时绕过 bug 的权宜之计，而是"我这次检索不需要 LLM 扩展查询"的明确声明。

> **【踩坑预警】**：关于 #1368 的状态要诚实——截至实验室 v0.42.44.0 实测，`--no-expand` 仍是必要的规避手段（不加它默认走 expansion 路径）。我们不去断言这个 Issue 在 GitHub 上是 open 还是 closed（那需要查官方状态），只陈述实测事实：在当前版本，生产查询默认加 `--no-expand` 是更稳妥的选择。至于硬超时——命令行直接 `!gbrain query` 没有 Python `subprocess.run(timeout=N)` 那种兜底，但只要带上 `--no-expand` 跳过 expansion 这一步，query 编码后直接走 HNSW + BM25 + RRF 立即返回，根本不会触发 #1368 的 CPU 挂死。所以 `--no-expand` 是根治手段，不是临时绕过。

&emsp;&emsp;**坑二：search 的检索路径取决于有没有 embedding key。** 这个坑 6.3 节已经埋过——`gbrain search` 默认走混合检索（带向量）；只有剔掉 embedding key、没了 provider，它才退成纯 BM25。做"BM25 vs 混合"对比实验时，这一步隔离不做，跑出来的数据就是错的。我们再正式跑一遍对照，把"隔离 vs 不隔离"的差异坐实。

In [29]:
# 对照实验：隔离 key（纯 BM25）——用 env -u 剔除 DASHSCOPE_API_KEY 强制纯 tsvector
# 对比：去掉 env -u 直接跑 gbrain search，同一 query 分数会落到约 1.0 的混合区间
!env -u DASHSCOPE_API_KEY gbrain search "NexaFlow" --limit 3

[0.4965] companies/nexaflow -- # NexaFlow
NexaFlow is a workflow startup founded in 2021 by [[people/alice-chen]].
The engineering 
[0.2918] people/carol-wu -- # Carol Wu
Carol Wu works at [[companies/nexaflow]] as CTO since 2022.
She manages engineering at Ne


&emsp;&emsp;这段代码用 `env -u` 剔除 `DASHSCOPE_API_KEY`，让 `gbrain search` 没有 embedding provider、退成纯 BM25。运行后你看到的分数（或无结果）才是真正的 BM25 一路数据。如果不剔除、让 key 留在环境里，同一条 `search` 默认走混合检索，分数会落到约 1.0 的 RRF 区间——那不是纯 BM25。这个隔离是做任何"BM25 vs 混合"对比实验的前提。下面不剔除 key 再跑一次，直接看到差异：

In [30]:
# 不剔除 key：同一 query，search 默认走混合检索（分数落到约 1.0 的 RRF 区间）
!gbrain search "NexaFlow" --limit 3

[1.1628] companies/nexaflow -- # NexaFlow
NexaFlow is a workflow startup founded in 2021 by [[people/alice-chen]].
The engineering 
[0.8688] people/carol-wu -- # Carol Wu
Carol Wu works at [[companies/nexaflow]] as CTO since 2022.
She manages engineering at Ne
[0.5150] people/alice-chen -- # Alice Chen
Alice Chen is the founder of [[companies/nexaflow]], which she co-founded in 2021.
Alic


&emsp;&emsp;把两次结果并排看：剔除 key 那次是纯 BM25（tsvector 量纲、偏高或无结果），这次带 key 是混合（分数落在约 1.0 的 RRF 区间）——同一条 `search` 命令，环境里有没有 embedding key，跑的是两条不同的检索路径。这正是 6.3 节强调"测 BM25 必须先隔离 key"的原因。

&emsp;&emsp;Retrieval 层补全了：你看清了向量、BM25、RRF 三路如何各补短板、融合后怎么重排，还避开了 #1368 和 DASHSCOPE 隔离两个工具坑。但到现在，检索给你的还只是"页面列表"——你得自己打开读。下一站，我们跨出最关键的一步：从"给你页面"到"给你答案"。

---

## <center>第 7 章：综合层（search vs think）</center>

&emsp;&emsp;这是本节课的重点。前面三站我们把存储和检索都铺好了，现在终于能回答第 1 章埋下的那个问题：前面只让你用了"建图"那半边，今天要补的"另一半"到底是什么。答案就在这一章——`search` 给你页面，`think` 给你答案。我们分两步走：先用 `search` 跑一遍，暴露"找到了页还得自己读"的局限，再请出 `think`，亲眼看它替你读完多页、综合成带引用的答案。前面第 1 章提到 `think` 输出末尾藏着最有价值的东西，这一章就来看它到底是什么。

### 7.1 search 的局限：只给页面

&emsp;&emsp;先看 search 的局限。设想一个真实场景：你明天要见 Alice，临睡前想问大脑"明天见 Alice 前我该知道什么"。我们先用 `search` 跑一下。

In [31]:
# Step 1：先用 search 跑一个真实场景问题——明天见 Alice 前该知道什么
!gbrain search "明天见 Alice 前我该知道什么" --limit 5

[0.8370] people/alice-chen -- # Alice Chen
Alice Chen is the founder of [[companies/nexaflow]], which she co-founded in 2021.
Alic
[0.8020] companies/sparkfield -- # SparkField
[[people/bob-morgan]] invested in SparkField as part of [[companies/alphaventures]]'s p
[0.7852] people/carol-wu -- # Carol Wu
Carol Wu works at [[companies/nexaflow]] as CTO since 2022.
She manages engineering at Ne
[0.7695] companies/nexaflow -- # NexaFlow
NexaFlow is a workflow startup founded in 2021 by [[people/alice-chen]].
The engineering 
[0.7486] people/bob-morgan -- # Bob Morgan
Bob Morgan works at [[companies/alphaventures]] as a general partner.
Bob Morgan invest


&emsp;&emsp;运行后你会看到 5 行结果，每行是 `[分数] slug -- 页面首句预览`，比如：

```
[0.8370] people/alice-chen -- # Alice Chen ...
[0.8020] companies/sparkfield -- # SparkField ...
[0.7852] people/carol-wu -- # Carol Wu ...
[0.7695] companies/nexaflow -- # NexaFlow ...
[0.7486] people/bob-morgan -- # Bob Morgan ...
```

&emsp;&emsp;**这就是 search 的局限。** `search` 确实找到了 5 个相关页面——Alice 本人、她当顾问的 SparkField、她公司的 CTO Carol、她创立的 NexaFlow、投资人 Bob。每一页都相关。但问题是：**你得自己打开这 5 页、一页页读完，再在脑子里把它们拼起来**，才能真正"准备好见 Alice"。这呼应了开头那句话——搜索给你页面，不等于给你答案。临睡前还要读 5 页做综合，这体验不行。

> **【30 秒小结】**：`search` 的产出是一个排好序的页面列表，综合工作量全留给了你。当你的问题需要"跨多页综合"才能回答时，光有检索是不够的——这正是 `think` 要补上的那一步。

### 7.2 think：多页综合成答案

&emsp;&emsp;下面用 `gbrain think` 做同一个问题。它是 GBrain 的端到端多跳合成命令，也是"从 search 到 think 跨越"的落点。它的工作方式是：先在 brain 里检索相关页面，再跨多个页面综合产出一段带 citation（引用标注）的答案，过程中还分析页面之间的冲突与空白。

&emsp;&emsp;调用 `think` 有一个关键姿势必须先讲清，否则会被拒：

> **【踩坑预警】**：`gbrain think` 默认走 Anthropic 的 Claude 模型，如果你想换成 DeepSeek，关键是**模型名必须带 provider 前缀**。直接写裸名 `gbrain think "..." --model deepseek-v4-flash` 会失败——GBrain 会把没有前缀的名字默认补成 `anthropic:deepseek-v4-flash`，于是在 Anthropic 名单里找不到它，报 `unknown_model`。正确写法是带上 `deepseek:` 前缀：`gbrain think "..." --model deepseek:deepseek-v4-flash`；或者用等价的环境变量前缀 `GBRAIN_MODEL="deepseek:deepseek-v4-flash" gbrain think "..."`——两种都有效，本质都是"给模型名补上 `deepseek:` 这个 provider 前缀"。`GBRAIN_MODEL` 能直接切底层 provider（Anthropic / OpenAI / DeepSeek），无需改任何配置文件——某个 provider 限额耗尽时换一个继续跑。注意：`--no-expand` 只对 `gbrain query` 有效（避开 6.7 节讲的 #1368），`gbrain think` 并没有这个标志，给 think 硬加会被当成问题文本的一部分、静默失效，所以 think 不要加 `--no-expand`。（**时效注**：DeepSeek 旧模型名 `deepseek-chat` / `deepseek-reasoner` 将于 2026-07-24 废弃，官方等价平替是 `deepseek-v4-flash`，本课已统一用 `deepseek:deepseek-v4-flash`；若要更强的 Pro 档，可换成 `deepseek:deepseek-v4-pro`。）

&emsp;&emsp;同一句话，我们这次改用 `think` 跑。注意 `think` 要调一次完整的 LLM 推理，会比 `search` 慢一些。

In [32]:
# Step 2：同一个问题，改用 think——替你读完多页、综合成带引用的答案
# 行内环境变量 GBRAIN_MODEL 切到 DeepSeek（避开默认 Anthropic 限额）
# 注意：think 没有 --no-expand 标志（那是 query 的），所以这里不加
!GBRAIN_MODEL=deepseek:deepseek-v4-flash gbrain think "明天见 Alice 前我该知道什么"

# 明天见 Alice 前我该知道什么

明天见 Alice Chen 前，以下是核心背景：

- Alice Chen 是 **NexaFlow** 的联合创始人，该公司成立于 2021 年，是一家工作流初创公司 [people/alice-chen][companies/nexaflow]。
- 她还担任 **SparkField** 的技术顾问，提供产品指导 [people/alice-chen][companies/sparkfield]。
- NexaFlow 的工程团队由 **Carol Wu** 领导，她是 CTO，与 Alice 在产品方面合作 [people/carol-wu]。
- 投资方面，**Bob Morgan**（AlphaVentures 的合伙人）参与了 NexaFlow 的种子轮并领投了 A 轮，同时也是 SparkField 的早期投资人 [people/bob-morgan][companies/nexaflow][companies/sparkfield][companies/alphaventures]。

这些关系网可能影响会谈内容，建议提前了解 NexaFlow 和 SparkField 的业务进展。

## Gaps
- Alice 的个人兴趣或当前业务重点
- NexaFlow 或 SparkField 的近期动态或融资情况

---
Model: deepseek:deepseek-v4-flash | Pages: 6 | Takes: 0 | Graph: 0 | Citations: 6


&emsp;&emsp;这一跑，输出就完全不是页面列表了，而是一段成文的中文答案。实验室真跑的结果是这样的：

```
## 与 Alice Chen 会面前的关键背景

### Alice Chen 的个人与职业身份
- Alice Chen 是工作流初创公司 NexaFlow 的创始人，公司成立于 2021 年[people/alice-chen]。
- 她还担任 SparkField 的技术顾问，提供产品指导[people/alice-chen][companies/sparkfield]。

### 她的核心人脉与关联方
- NexaFlow 的 CTO 是 Carol Wu（自 2022 年起），她管理工程团队并与 Alice 在产品上合作[people/carol-wu][companies/nexaflow]。
- AlphaVentures 是 NexaFlow 的 A 轮领投方，且合伙人 Bob Morgan 是种子轮投资者，并领导了 A 轮[people/bob-morgan][companies/nexaflow]。
- Bob Morgan 同样投资了 SparkField 的种子轮，因此 Alice 与 Bob 在两家公司都有交集[people/bob-morgan][companies/sparkfield]。

### 潜在话题
- 可以讨论 NexaFlow 的产品与工程（通过与 Carol Wu 的工作关系）。
- 可以提及 SparkField 的顾问角色，以及 AlphaVentures/Bob Morgan 在两家公司的参与。

---
**注意事项**：以上信息均来源于脑数据，未发现矛盾。

---
Model: deepseek:deepseek-v4-flash | Pages: 6 | Takes: 0 | Graph: 0 | Citations: 6
```

&emsp;&emsp;我们把它和 7.1 的 `search` 对比一下——这次你不用再读 5 页了。`think` 替你读完了所有相关页面，把散落的事实综合成一段结构化答案：Alice 是谁、她的核心人脉、明天能聊什么话题。而且**每句话后面都带着引用标注** `[people/alice-chen]`，你能一眼追溯每个结论出自哪一页。底部那行状态信息 `Pages: 6 | Citations: 6` 说明它这次检索到 6 页、产出了 6 条引用（这个样例库总共就 6 页都相关，真实大库里 think 只会取回与问题相关的那部分页，不是全库都读）。这就是从"给你页面"到"给你答案"的完整跨越——前面只覆盖了"给你页面"这一半，另一半在这里补齐了。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175016409.png" width=65%></div>

&emsp;&emsp;我们再看一遍 `think` 输出的内部流程，它做了四件事：**检索相关页 → 读多页 → 综合成文 → 标引用**。这四步里，前两步是 Retrieval 层（第 6 章）的活，后两步是 LLM 综合的活——所以你能理解为什么第 6 章是第 7 章的前提：`think` 内部就调着我们刚学的混合检索。

> **【踩坑预警】**：think 输出底部可能带一行 `Warnings: CITATIONS_REGEX_FALLBACK`。这是个提示性警告，表示 citation 是通过正则回退方式抽取的（不同 provider 的输出格式有差异，DeepSeek 的引用格式和默认的略有不同）。它不影响答案可用，记录在案即可，不用紧张——这是真实环境里会遇到的一个小工程坑，如实呈现给你。

### 7.3 空白分析：Gaps 段的价值

&emsp;&emsp;我们回到本章开头记下的那一笔——`think` 输出里"最容易被忽略却最有价值"的东西，就是**末尾的空白分析（gap analysis）**。

&emsp;&emsp;我们回看 7.2 那段输出的末尾：`**注意事项**：以上信息均来源于脑数据，未发现矛盾。` 这一句就是空白分析的雏形。在我们这个干净的小语料上，它说"未发现矛盾"；但在真实的、有时间跨度的大脑里，这段会变得极其有价值——它会诚实地告诉你"哪些信息可能过期、哪些结论没有引用支撑、哪些页面自相矛盾"。

&emsp;&emsp;为什么这才是 `think` 的差异化价值？因为正文那段综合答案，本质上是把已知信息整理一遍；而空白分析回答的是**"我还不知道什么"**——这是会前准备里最关键的部分。你最怕的不是"知道的信息组织得不好"，而是"以为自己准备好了、其实有个关键事实已经过期了"。下面这张实验室截图是一个更丰富的真实案例：一次跨 40 页、带 14 条引用的 `think`，末尾有完整的"知识盲区 Gaps"段。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175038633.png" width=80%></div>

&emsp;&emsp;我们看这张图里 think 输出的末尾，它逐条列出几类该警惕的信号——这里要分清：**缺失数据**列在 Gaps 段，**页面互相矛盾**列在单独的 Conflicts 段（prompt 里这是两条独立规则），而"信息过期"是模型读了页面的时间元数据后自己判断的。下面这张表把它们放一起对照：

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>think 输出末尾的几类警示信号</font></p>
<div class="center">

| 信号 | 含义 | 你该做什么 |
|------|------|-----------|
| 过期信息 | 某页数据已经很久没更新 | 去核实最新情况，别拿旧数据做决策 |
| 无引用论断 | 某个结论找不到支撑页面 | 警惕，可能是模型脑补的 |
| 自相矛盾 | 两页对同一事实说法不一致 | 人工判断哪个对，或两个都存疑 |

</div>

&emsp;&emsp;所以用 `think` 时，养成一个习惯：**先看末尾的空白分析，再看正文**。正文告诉你"已知的怎么样"，空白分析告诉你"未知的有哪些"——后者才是它真正帮你避坑的地方。

&emsp;&emsp;这里要澄清一个很容易搞错的问题：**空白分析到底是谁做的——是大模型自己思考出来的，还是 GBrain 的工程代码算出来的？** 很多人会以为这是 GBrain 用某种确定性规则（比如扫描时间戳找过期、检查每个论断有没有引用）自动算出来的——**不是这样**。

&emsp;&emsp;真相是：**空白分析的内容是大模型自己生成的**。`think` 在调用 LLM 时，发给模型的 prompt 里有一条硬性规则，明确要求它："如果大脑里没有相关数据导致你答不了，就在 **Gaps 段**里列出具体缺哪些信息，**不许编造答案**"（这是源码 `think/prompt.ts` 里的原文规则，实测确认）。模型读完检索到的那些页面之后，**自己判断**哪些信息缺失（写进 Gaps 段）、哪些页面互相矛盾（写进单独的 Conflicts 段），把结论写进它返回的 JSON。GBrain 的工程代码只做三件事：**写 prompt 提要求、解析模型返回的 JSON、把它渲染出来**——全程**没有任何"检测过期 / 检测缺引用 / 检测矛盾"的确定性算法**去计算 gaps 的内容。一个反证最能说明问题：如果你没配大模型（没有 API key），gaps 会直接退化成一句固定的占位话 `no LLM available; synthesis skipped`，根本给不出真正的盲区分析——没有 LLM，就没有空白分析。

&emsp;&emsp;所以准确的理解是：**空白分析是"大模型的思考 + GBrain 的约束框架"两者结合的产物**。盲区的具体内容是模型想出来的（它的思考过程），但"每次 `think` 都必须做盲区分析"这个行为，是 GBrain 用 prompt 硬性设计出来的强制动作。这个区分对你怎么用它很重要：① 盲区分析的质量**取决于你用的模型**——越强的模型，盲区找得越准、越全；② 它**不是 100% 确定性的**——同一个问题，换个模型、甚至同一模型跑两次，盲区列表可能略有出入。把它当成"一个聪明助手帮你查漏补缺的提醒"，而不是"一份机器算出来的精确清单"，心态就对了。

### 7.4 find_trajectory：时序综合

&emsp;&emsp;`think` 还有一个进阶能力值得知道——**find_trajectory（时序追踪）**。它能在一次查询里追出一个对象随时间的演变，比如"这个指标怎么变的、团队什么样了、当初承诺了什么"。

&emsp;&emsp;这个能力的价值在于"复利"：传统做法你得分别查指标、查团队、查承诺，再自己对齐时间线；find_trajectory 一次就把时序脉络综合出来。

&emsp;&emsp;先把"**它吃什么数据**"说精确，这是最容易误解的地方：find_trajectory 要的**不是"带时间戳的散文页"，而是 typed claims（带类型的结构化声明，GBrain 内部叫 `take`，就是 `think` 输出里那行 `Takes: N`）**。一条 take 是从页面文本里抽出来的结构化行，带几个关键字段：`kind`（类型，比如 `metric` 指标型、`event` 事件型）、`holder`（关于谁）、`since_date`（什么时候生效），还有"被新声明取代（superseded）"和"事后验证对错（resolved）"的状态。举个例子——"Acme 的 MRR 一季度 5 万、二季度涨到 8 万"这句散文，会被抽成两条 `metric` 型 take：`{holder: companies/acme, metric: MRR, value: 5万, since: Q1}` 和 `{…value: 8万, since: Q2}`；find_trajectory 正是把同一实体同一指标的这些 take 按时间排成曲线（5 万→8 万），还顺手算出涨跌（regression）和叙事漂移（drift）。所以**光有时间戳不够**，得是能被 schema 抽成 typed take 的结构化指标声明（这依赖第 8.3 章讲的 schema 定义了哪些 take kinds）。我们这 6 页 people/companies 是**静态实体页、没有随时间变化的 typed take**，所以演示不出来——这就是本节只点到概念的原因。

&emsp;&emsp;再说"**怎么触发**"，免得你以为非得记个特殊命令：find_trajectory 有**两种用法**。① **`think` 自动**——当你问的是时序类问题（"X 的指标怎么变的""当初关于 Y 说过什么"这类时间意图），`think` 默认就会自动抽出候选实体、跑 find_trajectory，把时序综合进答案，你什么都不用多做（想关掉用 `think.trajectory_enabled=false`）；② **独立的 `find_trajectory` 工具**——显式给一个 `entity_slug`（如 `companies/acme`），还能按 `metric`、时间范围（`since` / `until`）过滤，专门追某一个实体的指标曲线 + 自动检测回退和漂移。一句话：前者适合"答问题时顺带带出演变"，后者适合"精确盯住一个实体"。真正用它，要在装了真实 typed take 数据的大脑上。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174306010.png" width=65%></div>

### 7.5 回应第 1 章的问题

&emsp;&emsp;走到这里，第 1 章那个"另一半"的钩子完整闭合了。我们当时说：左边那台只会建图的大脑回答不了"我该准备什么"，因为它需要读完多页、综合成话、标清引用。现在你亲眼看到了——`think` 正是干这件事的：它在第 6 章的混合检索之上，加了"读多页 + LLM 综合 + 引用标注 + 空白分析"这一层，把"页面列表"变成了"会前准备好的答案"。

&emsp;&emsp;这就是前面没展开的那一半。从 `search` 到 `think`，从"给你页面"到"给你答案"——GBrain 作为完整 brain layer 的核心价值，在这一章落地了。但 `think` 不是没有代价的（它要先检索相关页、再跨页综合，还要调一次 LLM），这个代价我们留到第 9 章 benchmark 那里用数据算清。下面先去看看 GBrain 的工程生态。

---

## <center>第 8 章：操作与生态</center>

&emsp;&emsp;前面几章我们把 GBrain 的核心机制走完了。这一章退后一步看全景：GBrain 不只是几条检索命令，它有一整套生态——一堆 CLI 命令、51 个 skills、约 90 个 MCP 工具、可定制的 schema 包、还有一套夜间自动维护的 Minions 队列。这章不深挖每一个，而是帮你建立"这些东西分别是什么、彼此什么关系"的全景认知，并诚实标出几个版本边界。我们重点讲三件容易混淆的事：skills 和 MCP tools 为什么是两个数、schema 包机制、Minions 在 PGLite 上的边界。

### 8.1 CLI 命令速查

&emsp;&emsp;GBrain 的命令行覆盖了从建库到维护的全流程。这里给一张速查表，不逐条展开（多数你前面已经用过），知道有哪些、需要时回来查即可。

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>GBrain 常用 CLI 速查</font></p>
<div class="center">

| 类别 | 命令 | 作用 |
|------|------|------|
| 建库 | `init` / `migrate` / `doctor` | 建 brain / 引擎迁移 / 健康检查 |
| 页面 | `get` / `put` / `delete` / `list` | 读写删列页面 |
| 检索 | `search` / `query` / `ask` / `think` | 关键词 / 混合 / 问答 / 思考 |
| 导入导出 | `import` / `sync` / `export` | 导入目录 / 增量同步 / 导出 |
| 生态 | `schema` / `jobs` / `call` | schema 包 / 任务队列 / 调 skill |

</div>

&emsp;&emsp;这张表的价值是给你一个"地图"——前面我们用了 `init`/`import`/`search`/`query`/`think`/`schema`/`jobs`，它们都在这张图里有位置。其余命令需要时查 `gbrain --help` 即可，不占正课时间。

### 8.2 skills vs MCP tools：两个维度

&emsp;&emsp;接触 GBrain 生态你很快会撞上一个困惑：README 说有 43 个 skills、"30+" 个 tools，另一份梳理文档又说 74 个 tools，到底哪个对？答案是——**别信文档，直接用命令数**。文档更新永远追不上代码，活跃迭代的开源项目尤其如此。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174532449.png" width=55%></div>

&emsp;&emsp;实验室 case-5 在当时版本实测：**skills = 51**（README 的 43 已过时）、**MCP tools = 89**。下面这张实验室截图把三个口径并排打出来，让"文档过时、以实测为准"一目了然。要先说清楚的是：这个 **89 是 case-5 那次（v0.42.38.0）的数**，本机更新到 v0.42.44.0 后实测是 **90**——数字随小版本浮动，所以下文统一说"约 90 个 MCP tools"，下方的踩坑预警会展开这件事。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175040942.png" width=70%></div>

&emsp;&emsp;这两个数怎么对不上？因为它们测的根本不是一回事：

&emsp;&emsp;**skills 是指令层**——operator 手写的 prose 工作流文件，告诉 AI"做某件事的步骤是什么"（WHAT to do），它本身不是可执行代码。比如 `academic-verify` 这个 skill 是一段带触发词的文本，描述"如何核验一条学术主张"的工作流。

&emsp;&emsp;**MCP tools 是执行层**——可被调用的类型化 API，带明确的输入 schema，真正去读写页面、做搜索（HOW to do it）。比如 `get_page` 是一个带 `inputSchema` 的接口，真去执行"按 slug 读一个页面"。

&emsp;&emsp;两者的关系是"skills 调用 MCP tools 来完成操作"。而 GBrain 的配置项 `config.mcp.publish_skills=true` 还会把 skills 本身也暴露成 MCP tools，所以 MCP tools 数（约 90）会大于纯 CRUD 操作数。51 和约 90 不矛盾，是同一系统的两个切面。

> **【踩坑预警】**：这两个数字会随版本波动，别把它当死数。本机在 v0.42.44.0 实测 `gbrain --tools-json` 数出来是 **90 个** MCP tools，而 case-5 实验室在 v0.42.38.0 数的是 89——差了 1 个。这正是"小版本频繁变化"的又一例证。所以引用时说"约 90 个 MCP 工具、随版本浮动"，而不是钉死某个具体数。要拿你自己环境的真值，就跑 `gbrain call list_skills '{}'`（数 skills）和 `gbrain --tools-json`（数 tools），别查 README。

### 8.3 schema 包机制

&emsp;&emsp;GBrain 用 **schema 包**来定义"你的大脑认得哪些类型的知识"。默认包叫 `gbrain-base-v2`，它定义了一套页面类型（person、company、concept 等）和关系动词。我们真跑看一下当前激活的 schema。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175016344.png" width=55%></div>

In [33]:
# 查看当前激活的 schema 包：包名、类型数、关系动词数
!gbrain schema active 2>&1 | grep -iE 'Active pack|Page types|Link verbs|Takes kinds'

Active pack: gbrain-base-v2 v1.0.0
Page types: 15
Link verbs: 14
Takes kinds: fact, take, bet, hunch


&emsp;&emsp;输出会显示 `Active pack: gbrain-base-v2 v1.0.0`、`Page types: 15`、`Link verbs: 14`、`Takes kinds: fact, take, bet, hunch`。也就是说默认 schema 定义了 15 种页面类型、14 种关系动词。这套类型系统就是大脑"认知世界的框架"——它知道什么是 person、什么是 company、它们之间能有哪些关系。再看可用的 schema 包列表。

In [34]:
# 列出可用的 schema 包
!gbrain schema list

Bundled packs:
  gbrain-base
  gbrain-recommended

Installed packs (~/.gbrain/schema-packs/):
  my-domain


&emsp;&emsp;输出会列出 bundled 的包（`gbrain-base`、`gbrain-recommended`）。注意 `schema list` 看的是全局的 schema 包仓库（`~/.gbrain/schema-packs/`），如果你没装过自定义包，会提示 "No user-installed packs"——这是正常的，说明你用的是内置默认包。

&emsp;&emsp;schema 里还有一个命令 `gbrain schema detect`，本意是从你的文件夹结构聚类、建议类型。但它正好撞上一个版本 breaking，我们如实展示——这是一个"活教材"。

In [35]:
# schema detect：本意是从文件夹聚类建议类型，但 v0.42.44 上撞版本 breaking
# 2>&1 合并 stderr，直接看到报错：relation "pages" does not exist（版本 breaking 活教材）
!gbrain schema detect 2>&1

relation "pages" does not exist


> **【踩坑预警】**：在 v0.42.44.0 上，`gbrain schema detect` 会直接报 `relation "pages" does not exist`——表名不存在。这是个真实的版本 breaking：同一台机器从 0.42.38 升到 0.42.44 后，这条命令就挂了。我们不回避它，反而把它当成"版本不稳"的活教材给你看。应对办法是：在当前版本不用 `detect`，改用 `schema active` / `schema list` 来理解和管理 schema 包。这也再次印证了第 3 章"为什么要锁版本"那条预警——breaking change 是 GBrain 当前阶段的真实状态。

#### 8.3.1 动手：为你自己的领域设计一套 schema

&emsp;&emsp;内置的 `gbrain-base-v2` 是给"创投 / 科技"场景调好的（person、company、deal 这些类型）。如果你的知识库是别的领域——学术论文、法律案例、医疗记录——你会想定义自己的类型。这一节讲清楚怎么做，以及背后的方法论。

&emsp;&emsp;**先看清 schema 包的结构。** 上面 `schema active` 显示的"15 类型 / 14 动词 / 4 takes"，背后是三类定义：

&emsp;&emsp;① **页面类型（page_types）**——每个类型必须挂到 5 个"基元"之一：`entity`（实体，如 person/company）、`media`（媒介，如 tweet/分析）、`temporal`（时间，如 meeting）、`annotation`（标注）、`concept`（概念）。基元是**封闭枚举**（你只能新增"扩展某基元"的类型，不能造新基元），它决定这个类型的默认行为。每个类型还带两个**正交**的字段：`path_prefixes`（按目录前缀把 markdown 自动归类——`people/` 下的归 person）和 `aliases`（查询时的别名扩展——查 "founder" 也能命中 person）。

&emsp;&emsp;② **关系动词（link_types）**——如 `founded` / `founded_by`（成对的正反向边）、`mentions`（无反向）。

&emsp;&emsp;③ **断言种类（takes_kinds）**——`fact` / `take` / `bet` / `hunch`，对应"事实 / 观点 / 下注 / 直觉"四种认知颗粒。

&emsp;&emsp;**存在哪？** schema 包是 **YAML 文件，不是数据库表**——别搞混了（数据库表是 pages/links 那些，schema 包是"分类法"的声明）。内置包在 GBrain 源码里；你自定义的包放在 `~/.gbrain/schema-packs/<包名>/pack.yaml`；"当前用哪个包"记在 `~/.gbrain/config.json`。

&emsp;&emsp;**自定义流程（命令实测可跑）。** 上面试过的 `schema detect` 在当前版本是坏的，所以**别走自动检测那条路**，改用手写。第一步：用 `schema init` 脚手架一个空包。

In [36]:
# 脚手架一个新 schema 包（会在 ~/.gbrain/schema-packs/my-domain/ 生成 stub pack.yaml）
!rm -rf ~/.gbrain/schema-packs/my-domain   # 重跑先清掉旧的（init 遇目录已存在会报错退出）
!gbrain schema init my-domain 2>&1 | head -3

(experimental) Scaffolded pack `my-domain` at /Users/mac/.gbrain/schema-packs/my-domain/pack.yaml
Next: edit pack.yaml, then run `gbrain schema validate my-domain` and `gbrain schema use my-domain`.


&emsp;&emsp;它会打印 `(experimental) Scaffolded pack ...`，并生成一个 `pack.yaml` 骨架——`extends: gbrain-base` 表示在内置包基础上扩展，`page_types` / `link_types` 都是空的等你填。看一眼骨架长什么样。

In [38]:
# 看脚手架出的骨架结构
!cat ~/.gbrain/schema-packs/my-domain/pack.yaml

# Stub pack — extends gbrain-base by default. Add your own page_types below.
api_version: gbrain-schema-pack-v1
name: my-domain
version: 0.0.1
gbrain_min_version: 0.39.0
extends: gbrain-base
description: "Stub pack scaffolded by 'gbrain schema init my-domain'. Edit /Users/mac/.gbrain/schema-packs/my-domain/pack.yaml then 'gbrain schema validate' + 'gbrain schema use my-domain'."

page_types:
  - name: paper                 # 新类型：论文
    primitive: media            # 挂到 media 基元（论文是一种媒介）
    path_prefixes: [papers/]    # papers/ 目录下的 md 自动归为 paper
    aliases: [arxiv, preprint]  # 查询时这些词也能匹配到 paper
    extractable: true           # 这个类型参与事实自动抽取
link_types:
  - name: cites                 # 新关系：引用
    inverse: cited_by           # 反向边：被引用
takes_kinds:
  - fact
  - take
  - bet
  - hunch
borrow_from: []


&emsp;&emsp;接下来**用编辑器编辑这个 pack.yaml**，按你的领域填类型和关系。下面是一个"学术论文"领域的最小示例（把它写进 pack.yaml 的 `page_types` / `link_types` 段）：

```yaml
# pack.yaml 片段示例：学术论文领域
page_types:
  - name: paper                 # 新类型：论文
    primitive: media            # 挂到 media 基元（论文是一种媒介）
    path_prefixes: [papers/]    # papers/ 目录下的 md 自动归为 paper
    aliases: [arxiv, preprint]  # 查询时这些词也能匹配到 paper
    extractable: true           # 这个类型参与事实自动抽取
link_types:
  - name: cites                 # 新关系：引用
    inverse: cited_by           # 反向边：被引用
```

&emsp;&emsp;填好后，先**校验格式**（GBrain 用 Zod schema 把关字段对不对、基元是不是合法枚举、版本号是不是 semver），再**激活**它。

In [39]:
# 校验包格式（这里校验的是还没填类型的骨架，所以 page types 显示 0；
# 你按上面示例填好 paper/cites 后再跑，会看到对应的类型数 / 动词数）
!gbrain schema validate my-domain 2>&1 | head -5

✓ my-domain v0.0.1: valid manifest
  Path: /Users/mac/.gbrain/schema-packs/my-domain/pack.yaml
  Page types: 1
  Link verbs: 1
  Takes kinds: 4


&emsp;&emsp;校验通过会打印 `valid manifest`（这里因为还没填类型，page types 显示 0；填好 paper/cites 后再跑会看到对应数量）。最后一步是激活——但它会切换全局默认 schema，所以演示阶段你可以先只验证、不激活。

In [ ]:
# 激活这个包 —— 注意：这会把全局默认 schema 切成 my-domain，影响你之后新建的库！
# 演示阶段建议先不激活、只验证格式；真要用时再取消注释。
# !gbrain schema use my-domain
# !gbrain schema active   # 确认切换成功，并看是哪一层设置的
print("schema use 会切换全局默认包，按需取消注释运行；不确定就先只 validate")

> **【踩坑预警】**：① 想用 `schema fork gbrain-base-v2 <新名>` 从内置包复制一份来改？实测会报 `Source pack gbrain-base-v2 not found`——**fork 对 bundled 内置包不可用**。老老实实用 `schema init`（它已经 `extends gbrain-base`，等于站在内置包的肩膀上扩展）。② `schema init` / `fork` 在源码里标了 `(experimental)`，是实验性命令，心里有数。③ 编辑包后**一定先 `validate` 再 `use`**——格式错了 `use` 会失败，先校验能省掉来回试错。

In [40]:
!cat ~/Git/gbrain-master/src/core/schema-pack/base/gbrain-base-v2.yaml

# gbrain-base-v2 — 15-type DRY/MECE canonical taxonomy
#
# v0.42 (T22) — the response to issue #1479. A production gbrain brain
# (186K pages) had accreted 94 distinct `pages.type` values in 9 clusters
# of redundancy. This pack collapses them to 14 canonical types plus
# `note` as the catch-all (15 total) using subtypes/aliases/mapping_rules.
#
# STANDALONE pack (D11/F1): NO `extends:` field. Declaring inheritance
# would compose gbrain-base's 24 types on top of the 15 declared here,
# breaking the "14 canonical types" claim. Old types are reached via
# the alias graph (query closure) plus the catch-all retype rule (data
# migration).
#
# Pack-upgrade declaration (D7): `migration_from: {pack: gbrain-base,
# version: 1.x}` fires the `pack_upgrade_available` onboard check for
# brains on gbrain-base@1.0.0 → 1.x.x.
#
# Mapping rules (D11+D12) drive the unify-types Minion handler:
#   - retype: change pages.type from from_type to to_type (with optional
#     subtype JSONB stamp + legacy_t

&emsp;&emsp;**设计一套 schema 的通用方法论**（这套顺序是 `gbrain-base-v2` 包自身的设计注释总结出来的，可直接套用）：

&emsp;&emsp;**第一步，先列页面类型、每个挂一个基元。** 问自己："我的知识里有哪些'东西'？每个属于实体 / 媒介 / 时间 / 标注 / 概念里的哪一种？"

&emsp;&emsp;**第二步，狠狠收敛类型数量（DRY / MECE）。** `gbrain-base-v2` 的注释记了个真实教训：一个 18 万页的生产大脑曾累积出 94 个混乱的 type 值、9 个冗余簇，最后收敛到 14+1 个规范类型。**宁可少类型 + 用 subtype 细分，也不要类型爆炸。**

&emsp;&emsp;**第三步，path_prefixes 和 aliases 分开想。** 前缀决定"这页归哪个类型"，别名决定"查询能不能命中它"——两件正交的事，别混为一谈。

&emsp;&emsp;**第四步，再定关系动词、成对声明 inverse。** 如 `cites` / `cited_by`、`authored` / `authored_by`；不需要反向语义的（如 `mentions`）就用单名。

&emsp;&emsp;**第五步，标好两个开关。** `extractable`（这个类型要不要自动抽事实）、`expert_routing`（要不要参与"谁懂这个领域"的专家查询）。

&emsp;&emsp;**第六步，老数据用 mapping_rules 迁移。** 如果你已经有一堆旧类型的页面，写迁移规则把旧类型映射到新类型，不用手动逐页重标。

&emsp;&emsp;一句话收口：**schema 的本质是教大脑"用你的语言认识你的知识"**——内置包替你想好了创投场景，换个领域，你就是给大脑重新定义一套"它该认得哪些东西、东西之间有哪些关系"。

### 8.4 Minions 队列与边界

&emsp;&emsp;最后一块生态是 **Minions 任务队列**。它是 GBrain 的后台任务系统，设计上支持"梦循环"（dream cycle）——夜里自动给大脑做维护：去重、修引用、检测矛盾。它还有崩溃安全的两阶段持久化，任务跑一半崩了也不丢。我们真跑看一下队列状态。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175016322.png" width=55%></div>

In [41]:
# 查看 Minions 任务队列健康状态
!gbrain jobs stats

Job Stats (last 24h):
  No jobs in the last 24 hours.

  Queue health: 0 waiting, 0 active, 0 stalled
  Lease pressure (1h): 0 bounces


&emsp;&emsp;输出会是一个健康面板：`No jobs in the last 24 hours.`、`Queue health: 0 waiting, 0 active, 0 stalled`——我们的新库还没提交过任务，队列是空的、健康的。再试着提交一个任务（用 `--dry-run` 只演示不真跑）。

In [42]:
# 提交一个 sync 任务，--dry-run 只确认会提交什么、不真执行
!gbrain jobs submit sync --dry-run

[DRY RUN] Would submit job:
  Name: sync
  Queue: default
  Priority: 0
  Max attempts: 3
  Data: {}


&emsp;&emsp;输出会显示 `[DRY RUN] Would submit job:`，列出任务名、队列、优先级、最大重试次数。但要先厘清一个最容易误解的点：**`--dry-run` 只是「预演」**——它把这些参数打印出来就直接返回了，**连入队都不做**，纯粹给你确认"这个任务长什么样"，不产生任何实际效果。

&emsp;&emsp;那"真正提交并执行"是怎么走的？关键认知是：**`submit` 只负责「入队」，不负责「执行」**。一条 `gbrain jobs submit sync`（去掉 `--dry-run`）做的事，是把任务塞进 `minion_jobs` 表、状态标成 `waiting`，然后就结束了——任务在队列里排着，等人来跑。真正把它捞出来执行的是另一个角色：**worker 进程**。整条链路是三段式：

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>Minions 任务执行的三段式链路</font></p>
<div class="center">

| 阶段 | 命令 | 干什么 | 谁来做 |
|------|------|--------|--------|
| ① 入队 | `gbrain jobs submit <name>` | 把任务写进 `minion_jobs` 表，状态 `waiting` | 你（提交方） |
| ② 消费 | `gbrain jobs work` | 启动 worker，**每 5 秒轮询**队列，捞出 `waiting` 任务执行，直到 Ctrl-C | 常驻 worker 进程 |
| ③ 执行 | worker 内部调用对应 handler | 真正跑 `sync` / `embed` / `import` 等逻辑 | worker |

</div>

&emsp;&emsp;两个关键认知补一下。**第一，worker 是「轮询」不是「定时」。** `gbrain jobs work` 启动后进入一个循环，默认每 **5 秒**扫一次队列、把 `waiting` 任务捞出来跑，直到收到中断信号。它**不是 cron**——GBrain 没有内置"每天几点自动触发"的调度器；没有 worker 在跑，submit 进去的任务就一直躺着不动。想周期性地跑（比如每小时 sync 一次），得靠**系统的 crontab** 从外部驱动。`autopilot-cycle` 这类任务也只是"跑一轮"（sync+extract+embed+backlinks），不会自己循环。

&emsp;&emsp;**第二，`--dry-run` 和 `--follow` 是两个极端。** `--dry-run` 是预演（不落地）；`--follow` 则相反——提交后**在当前进程内联执行**（源码输出 `Executing inline...`），不依赖外部 worker、当场跑完给你结果。这个 flag 对我们用 PGLite 的人尤其关键，原因就是下面这条边界：

> **【踩坑预警 · PGLite 上怎么真正跑任务】**：常驻 worker daemon `gbrain jobs work` **在 PGLite 上跑不起来**——不是"需要换成 Postgres 引擎"，真正的原因是 PGLite 用**独占文件锁**（单进程单连接），一个常驻 worker 会和你的主进程抢锁（官方文档原话：`gbrain jobs work` does not run on PGLite, exclusive file lock）。但要分清：**队列机制本身 PGLite 是支持的**——`jobs submit` / `jobs stats` / `--dry-run` 全都正常，跑不起来的只是那个"在后台持续轮询"的常驻 worker。那 PGLite 上想真正执行一个任务怎么办？**用 `gbrain jobs submit <name> --follow`，内联执行、当前进程直接把任务跑完**（如 `gbrain jobs submit sync --follow`）。想要"夜间自动维护"的效果，就把带 `--follow` 的命令写进系统 crontab 定时调用。一句话收口：真正的常驻 worker daemon「梦循环」是 Postgres 团队大脑的能力（呼应 5.3 节"PGLite 不支持后台 worker"那条）；PGLite 个人大脑靠 `--follow` 内联 + 系统 cron 来替代。这不是 bug，是 PGLite 单进程架构的固有边界。

#### 8.4.1 进阶实操：在 Postgres 上真跑一次 worker（需 docker pg）

&emsp;&emsp;上面这条边界讲的是"PGLite 跑不了常驻 worker"。但 worker 到底长什么样、三段式怎么流转，光看文字不如真跑一遍。这一节我们**临时借用 5.3 节启动的 Docker Postgres 容器**（`gbrain-pg`），把活动引擎切到 Postgres、真正起一个 worker 消费一个任务，把"submit 入队 → work 消费 → 执行完成"完整走通。

> **【运行前提与隔离声明】**：① 需要 5.3 节的 Docker Postgres 容器 `gbrain-pg` 正在运行（`docker ps` 能看到、5432 端口通）；② 本节**全程隔离**——新建一个独立的 `gbrain_jobs_demo` 库做演示，**不碰**你的 PGLite 主库 `./output/brain-l2`，也不碰 `postgres` 默认库；③ 演示前自动备份 `config.json`，**最后一个 cell 会把引擎切回 PGLite 并删掉 demo 库**，环境完整复原。所以下面这几个 cell **必须从上到下连续跑完**，尤其别跳过最后的清理 cell。

&emsp;&emsp;**第 ① 步：备份配置 + 建隔离库 + 切到 Postgres 引擎。** 我们复用 5.3 节配好的 Postgres 连接，只把库名换成隔离的 `gbrain_jobs_demo`（连接串经环境变量传递，密码不进命令行）。

In [43]:
# 进阶实操起手：备份当前配置、建独立 demo 库、把活动引擎切到 Postgres
import json, os, subprocess
from urllib.parse import urlparse, urlunparse

CFG = os.path.expanduser("~/.gbrain/config.json")
subprocess.run(["cp", CFG, CFG + ".demobak"], check=True)   # 备份，最后一个 cell 据此还原回 PGLite

# 复用 5.3 节配好的 Postgres 连接，仅把库名换成隔离的 gbrain_jobs_demo
base = urlparse(json.load(open(CFG))["database_url"])
os.environ["GBRAIN_DEMO_URL"] = urlunparse(base._replace(path="/gbrain_jobs_demo"))  # 经 env 传递，不暴露密码
PG_USER = base.username                                       # 容器超级用户（5.3 节设的，通常是 postgres）

# 建隔离 demo 库（已存在则复用），再 init 切到 Postgres（--no-embedding：演示队列不需要向量）
subprocess.run(f"docker exec gbrain-pg createdb -U {PG_USER} gbrain_jobs_demo", shell=True)
r = subprocess.run('gbrain init --url "$GBRAIN_DEMO_URL" --no-embedding --force --non-interactive',
                   shell=True, capture_output=True, text=True)
print("\n".join(l for l in (r.stdout + r.stderr).splitlines()
                if any(k in l for k in ("Engine", "Schema version", "ready", "Brain"))))
print("当前引擎 →", json.load(open(CFG))["engine"])           # 应为 postgres

Brain ready. 0 pages. Engine: Postgres (Supabase).
--- GBrain Mod Status ---
  Schema version 1 → 117 (112 migration(s) pending)
      Brain-augmented web research. Sends brain context to Perplexity so
      the search focuses on what is NEW vs already-known.
当前引擎 → postgres


&emsp;&emsp;输出会显示 `Brain ready. ... Engine: Postgres`、一长串 `Schema version 1 → 117` 迁移，以及末行 `当前引擎 → postgres`——活动库已经从 PGLite 切到隔离的 Postgres demo 库了。

&emsp;&emsp;**第 ② 步：派一个任务进队列，确认它在 `waiting` 排队。** 我们用一个 `shell` 任务（跑一句 `echo`）当演示载体——它不依赖任何页面数据，最干净地展示队列流转。`shell` 任务要求 worker 端带 `GBRAIN_ALLOW_SHELL_JOBS=1` 才会被执行。

In [44]:
# 派一个 shell 任务（echo）；此刻还没 worker，它会停在 waiting
!GBRAIN_ALLOW_SHELL_JOBS=1 gbrain jobs submit shell --params '{"cmd":"echo hello-from-minions-worker","cwd":"/tmp"}'
!gbrain jobs stats


⚠  Shell jobs require GBRAIN_ALLOW_SHELL_JOBS=1 on the worker process.
   Your job was queued (id=1) but will sit in 'waiting' until a
   worker with the env flag starts. To run now:

     GBRAIN_ALLOW_SHELL_JOBS=1 gbrain jobs submit shell \
       --params '...' --follow

   Or start a persistent worker (Postgres only — PGLite uses --follow):

     GBRAIN_ALLOW_SHELL_JOBS=1 gbrain jobs work

{
  "id": 1,
  "name": "shell",
  "queue": "default",
  "status": "waiting",
  "priority": 0,
  "data": {
    "cmd": "echo hello-from-minions-worker",
    "cwd": "/tmp"
  },
  "max_attempts": 3,
  "attempts_made": 0,
  "attempts_started": 0,
  "backoff_type": "exponential",
  "backoff_delay": 1000,
  "backoff_jitter": 0.2,
  "stalled_counter": 0,
  "max_stalled": 5,
  "lock_token": null,
  "lock_until": null,
  "delay_until": null,
  "parent_job_id": null,
  "on_child_fail": "fail_parent",
  "tokens_input": 0,
  "tokens_output": 0,
  "tokens_cache_read": 0,
  "depth": 0,
  "max_children": null,
 

&emsp;&emsp;`jobs stats` 会显示 `Queue health: 1 waiting, 0 active`——任务进队列了，但因为还没有 worker 在跑，它就停在 `waiting`（可能附一条 "WEDGED QUEUE" 提示，意思正是"有活在排队、却没人消费"，符合预期）。

&emsp;&emsp;**第 ③ 步：启动 worker 消费这个任务。** `gbrain jobs work` 是常驻轮询进程（每 5 秒扫一次队列），正常会一直跑到你 Ctrl-C。在 Notebook 里我们用 `subprocess` 给它设一个 22 秒超时——足够它启动、轮询、消费掉任务后自动停下，不会把 cell 卡死。

In [45]:
# 启动 worker 消费队列；22s 超时自动停（否则常驻进程会一直占着 cell）
import subprocess, os
env = {**os.environ, "GBRAIN_ALLOW_SHELL_JOBS": "1"}
try:
    subprocess.run(["gbrain", "jobs", "work"], env=env, timeout=22)
except subprocess.TimeoutExpired:
    print("\n[worker 已运行 22s，任务消费完毕，演示自动停止]")

[minion worker] shell handler enabled (GBRAIN_ALLOW_SHELL_JOBS=1)
[minion worker] subagent handlers enabled
[minion worker] brain-health-100 handlers registered (12 ops, 4 protected) + embed-backfill (v0.40) + embed-catch-up (v0.42) + unify-types (v0.42) + skillopt (v0.42.0.0, protected)


Minion worker started (queue: default, concurrency: 1, watchdog: 16384MB (auto-sized from 64GB host RAM), health-check: 60s)
Registered handlers: sync, embed, lint, extract-conversation-facts, enrich, lint-fix, integrity-auto, sync-retry-failed, import, extract, backlinks, contextual_reindex_per_chunk, autopilot-cycle, shell, subagent, subagent_aggregator, ingest_capture, reindex, repair-jsonb, orphans, integrity, purge, synthesize, patterns, consolidate, extract_facts, resolve_symbol_edges, recompute_emotional_weight, extract-atoms-drain, embed-backfill, extract-ner, extract-takes-from-pages, extract-timeline-from-meetings, embed-catch-up, unify-types, skillopt

[worker 已运行 22s，任务消费完毕，演示自动停止]


&emsp;&emsp;你会看到 worker 的启动日志：`Minion worker started (queue: default, concurrency: 1 ...)` 以及它注册的一长串 handler（含 `shell`）。它在首次轮询时 claim 到我们那个任务、执行、完成。22 秒后超时提示打印，worker 停下。

&emsp;&emsp;**验证执行结果。**

In [46]:
# 看任务详情：应为 COMPLETED，Result 里有 echo 的输出
!gbrain jobs get 1
!gbrain jobs stats

Job #1: shell (COMPLETED)
  Queue: default | Priority: 0
  Attempts: 0/3 (started: 1, stalled: 0/5)
  Backoff: exponential 1000ms (jitter: 0.2)
  Started: 2026-06-24T13:04:54.873Z
  Finished: 2026-06-24T13:04:54.884Z
  Result: {"pid":53878,"exit_code":0,"duration_ms":6,"stderr_tail":"","stdout_tail":"hello-from-minions-worker\n"}
  Data: {"cmd":"echo hello-from-minions-worker","cwd":"/tmp"}
Job Stats (last 24h):
  Type           Total   Done    Failed   Dead   Avg Time
  shell          1       1       0        0      0.0s

  Queue health: 0 waiting, 0 active, 0 stalled
  Lease pressure (1h): 0 bounces


&emsp;&emsp;`jobs get 1` 会显示 `Job #1: shell (COMPLETED)`，`Result` 里 `exit_code=0`、`stdout_tail="hello-from-minions-worker\n"`——echo 真被 worker 执行了；`jobs stats` 的 `Done` 变成 1、队列回到 `0 waiting`。这就把"submit 入队 → work 消费 → 执行完成"的三段式在真实 worker 上跑通了，也实证了上面那条边界的另一面：**Postgres 引擎下常驻 worker 确实能起来**。

&emsp;&emsp;**第 ④ 步：清理还原（务必执行）。** 把引擎切回 PGLite 主库、删掉隔离 demo 库，让环境回到演示前的状态。

In [47]:
# 清理还原：恢复 config（切回 PGLite brain-l2）+ 删 demo 库；容器保持运行
import os, json, subprocess
CFG = os.path.expanduser("~/.gbrain/config.json")
subprocess.run(["cp", CFG + ".demobak", CFG], check=True)                             # 还原配置
os.remove(CFG + ".demobak")
subprocess.run("docker exec gbrain-pg dropdb -U postgres gbrain_jobs_demo", shell=True)  # 删隔离库
print("已还原，当前引擎 →", json.load(open(CFG))["engine"], "| 库 →", json.load(open(CFG))["database_path"])

已还原，当前引擎 → pglite | 库 → ./output/brain-l2


&emsp;&emsp;末行打印 `当前引擎 → pglite | 库 → ./output/brain-l2`，说明环境已完整复原（容器 `gbrain-pg` 仍在运行，方便你继续测试）。最后回到主线收个口：上面这套是 **Postgres 团队大脑**才能做的常驻 worker 演示；回到你日常用的 **PGLite 个人大脑**，等价的"真正执行一个任务"靠的是上一节说的 `--follow` 内联（如 `gbrain jobs submit sync --follow`），当前进程直接把任务跑完，不需要常驻 worker。

&emsp;&emsp;生态全景到这里：你建立了"CLI 全貌、skills vs tools 两维度、schema 包、Minions 队列"的整体认知，也看清了几个版本边界和 PGLite 的能力天花板。还剩最后一块拼图——这东西到底行不行？这需要 benchmark 数字，也需要看懂这些数字的诚实口径。下一章我们就把判断框架立起来。

---

## <center>第 9 章：Benchmark 评测体系</center>

&emsp;&emsp;面对一个"正在火"的技术，你最该有的判断力是：先问数据，再下结论。GBrain 有公开的 benchmark 分数，但这些数字很容易被误读、被混引。这一章我们分两步走：先以理解的视角搞清"benchmark 到底测什么"，再以评判的视角看清两个 benchmark 的可信度差异。最后把它和"21× token 成本"的话题接起来，讲清一件容易混淆的事——成本贵不等于质量差。

### 9.1 benchmark 测什么

&emsp;&emsp;先把视角放在"理解"上。检索系统的 benchmark，测的无非是"召回得准不准、综合得好不好"。GBrain 主要引用的公开 benchmark 是 **LongMemEval**。

&emsp;&emsp;LongMemEval 是一个公开 benchmark，语料托管在 HuggingFace，由 500 个来自真实个人助理场景的问题组成。它的评判方式是 **keyword-only**（纯关键词匹配，不用 LLM 当裁判），所以**完全可复现**——任何人下载语料、跑 eval，都能得到一致的分数。GBrain 在它上面的分数是 **R@5 = 97.6%**。

&emsp;&emsp;R@5 是 **Recall@5**（前 5 个召回结果的召回率）的缩写——意思是"该找到的相关内容，有多少落在了返回的前 5 个里"。97.6% 是个相当高的检索召回率。这个数字的来源要标清楚：

> **【数字口径】**：LongMemEval R@5 = 97.6% 引自官方评测仓库 gbrain-evals 的公开发布（截至 v0.40.6.0 的 benchmark snapshot）。这是公开可复现的数字——LongMemEval 语料公开、评判 keyword-only 无 LLM 裁判，任何人都能复跑验证。第三方评测（如 vectorize.io）也独立确认了这个数字。所以引用它时可以直接用，标注"LongMemEval R@5 97.6%，公开可复现"即可。

### 9.2 BrainBench 与它的 caveat

&emsp;&emsp;GBrain 还有第二个 benchmark 分数——**BrainBench P@5 = 49.1%**。把它和 LongMemEval 的 97.6% 放一起，很容易得出一个错误印象："GBrain 在一个测试上接近满分、在另一个测试上只有一半，是不是不靠谱？"

&emsp;&emsp;这个对比是**无效的**，因为两个 benchmark 测的根本不是同一件事。P@5 是 **Precision@5**（前 5 个结果的精确率）。BrainBench 这个数字背后有几个必须知道的附加前提（caveat，即"引用时必须一起说明的限定条件"）：

> **【数字口径与 caveat】**：BrainBench P@5 = 49.1% / R@5 = 97.9%，引自 gbrain-evals 官方发布。但它有三个关键 caveat 必须随数字一起讲：① **厂商自建语料**——它不是公开数据集，是 GBrain 团队自己造的一个虚构世界（公司叫 Acme、Helix Labs，人物、会议全是合成的，约 240 个虚构 JSON 实体，官方亦记作 241）；② **用 LLM 当裁判**（不像 LongMemEval 的 keyword-only），评判带主观性；③ **样本规模**约 145 个查询 / 261 个子问题。所以这个 49.1% 只在"这个特定虚构世界、用 LLM 评判"的语境下成立，绝不能当作"GBrain 真实场景只能做到一半"来解读。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624175042796.png" width=70%></div>

&emsp;&emsp;注意这张图里两根柱子的视觉处理是刻意的：LongMemEval 用蓝色实心、BrainBench 用橙色斜纹，从视觉上就提示"这是两种不同性质的东西"，而不是同一把尺子上的两个刻度。

### 9.3 两个 benchmark 严禁混引

&emsp;&emsp;把两者的底层差异列清楚，"不可混引"就不是一句空话而是有据可依。

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>LongMemEval vs BrainBench 关键差异</font></p>
<div class="center">

| 维度 | LongMemEval | BrainBench |
|------|-------------|------------|
| 语料来源 | 公开（HuggingFace） | 厂商自建虚构世界 |
| 实体性质 | 真实个人助理 500 问 | 约 240 个虚构 JSON 实体 |
| 评估方法 | keyword-only（可复现） | LLM judge（带主观） |
| 可重现性 | 公开，人人可复跑 | 依赖官方提供 |
| 分数含义 | 通用召回能力 | 虚构世界里的检索能力 |
| 引用时 | 直接引用即可 | 必须注明"自建语料、虚构世界" |

</div>

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174228811.png" width=55%></div>

&emsp;&emsp;关于第三方交叉佐证，我们这里要做一次诚实的处理。在更广的内存系统评测里，有一个独立项目 **MemPalace** 也报告过 LongMemEval R@5 = 96.6% 的成绩，可以作为同 dataset 的横向参照。但引用它时必须加一个 caveat：

> **【数字口径与 caveat】**：MemPalace 96.6% 这个数字真实存在（已被 arXiv 论文独立复现），但它实际上是 **MemPalace 用 ChromaDB 默认 embedding 跑出来的裸向量检索结果**，并非 MemPalace "宫殿架构"本身带来的增益（这一点在该项目的 GitHub issue 里有争议讨论）。所以引用它时要说清"这是裸向量检索的基线，架构增益存争议"，而不能简单说成"MemPalace 架构达到 96.6%"。诚实地讲，把它当作"同 dataset 上 GBrain 97.6% 略高于 MemPalace 96.6% 基线"的横向参照即可，不要过度解读。

&emsp;&emsp;你可能注意到我们没有引用某些网上流传的其他第三方数字——那是因为有些数字（比如曾出现在一些汇总里的所谓"Schift 96.0%"）经过联网核查找不到任何可溯源的来源，无法验证。按"每个数字都要能说清出处"的纪律，查不到来源的数字我们一律不引用，免得拿去用的时候被人一问出处就露馅。这本身就是 benchmark 引用的一条铁律：**宁可少引一个数，不引一个查无实据的数**。

### 9.4 成本贵 ≠ 质量差

&emsp;&emsp;最后我们把 benchmark 和"21× token 成本"接起来，因为这两件事最容易被混为一谈。

&emsp;&emsp;你可能听过一个数字：编译式 RAG 每次查询的 token 消耗约是向量 RAG 的 21 倍。这个数字来自一篇预注册实验论文（arXiv 2605.18490），结论方向是确凿的——编译式查询确实更贵。

> **【数字口径】**：21× token 成本来自 arXiv 2605.18490 的预注册实验。引用时要加一个小样本注：该实验只有 **13 个问题 / 24 篇论文**的小语料（作者自己标注为 "small multi-domain corpus"），所以是"这个小样本下的预注册发现"，不能当作普适定律。方向（编译式更贵）是可靠的，但精确的"21 倍"在更大语料上可能浮动。

&emsp;&emsp;关键是要分清两件不同的事：**"成本贵"和"质量差"是两个独立的维度**。

&emsp;&emsp;编译式 RAG 贵（21× token），是因为 `think` 要做多跳综合——先检索相关页、跨页推理、再调 LLM 生成带引用的答案，回想第 7 章那个 `Pages: 6` 就是它这次实际取用的页数；问题越复杂、相关页越多，读得和算得就越多。但贵不代表它质量差：LongMemEval 97.6% 的高召回说明它检索质量很好；它的真正优势在跨源综合质量、可读知识库、可溯源引用——这些是花更高成本换来的**别的东西**，不是用更高成本换更低成本。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174229765.png" width=55%></div>

&emsp;&emsp;所以一个完整的判断要拆成三维来看，用一张表收尾本章。

<style>
.center {
width: auto;
display: table;
margin-left: auto;
margin-right: auto;
}
</style>
<p align="center"><font face="黑体" size=4>GBrain 三维权衡</font></p>
<div class="center">

| 维度 | GBrain（编译式）表现 | 怎么看 |
|------|---------------------|--------|
| 检索质量 | 高（LongMemEval R@5 97.6%） | 公开可复现，可信 |
| 查询成本 | 高（约 21× token，小样本注） | 架构决定，跨源综合的代价 |
| 适用场景 | 稳定知识库 + 需深度综合 | 高频变更场景向量 RAG 更划算 |

</div>

&emsp;&emsp;这张表就是判断 GBrain 适不适合你的框架：如果你的知识库稳定、且真需要跨源综合的深度，21× 成本花得值；如果是高频变更、只需简单检索，那向量 RAG 仍是更合理的选择。看懂数字的诚实口径、不混引、不把成本和质量混为一谈——这就是 benchmark 这一章要给你的判断力。下面进入最后一章，盘点你这节课带走了什么。

---

## <center>第 10 章：本节小结</center>

&emsp;&emsp;一节课走完，我们从"GBrain 只用了一半"出发，把另一半补全了。这一章我们做两件事：回顾五站走过的路、盘点你现在能做什么，最后诚实标出本节没深入的边界。

### 10.1 全程回顾

&emsp;&emsp;开篇我们说要走五站补全"另一半"，现在我们每站收一句话。

<div align=center><img src="https://typora-photo1220.oss-cn-beijing.aliyuncs.com/DataAnalysis/ZhiJie/20260624174312920.png" width=55%></div>

&emsp;&emsp;**架构全景**：我们建立了 brain ⊥ source 两轴的心智模型——git markdown 是源真相，数据库是编译产物，改了源不会自动同步、得重 import。

&emsp;&emsp;**Storage 层**：我们补全了存储原理——PGLite 和 Postgres 两引擎一契约、第一次把 embedding 开起来、避开了 `npm install` 抢注的三道装机坑。

&emsp;&emsp;**Retrieval 层**：我们看清了混合检索——向量补语义、BM25 抓精确词、RRF 按名次融合（不能直接加分，因为量纲不可比），外加 #1368 和 DASHSCOPE 隔离两个工具坑。

&emsp;&emsp;**综合层**：我们跨出了最关键一步——`search` 给页面、`think` 给答案，think 替你读完多页综合成带引用的文，末尾还有诚实的空白分析。这是本节的核心高峰。

&emsp;&emsp;**生态与评判**：我们建立了全景认知——51 skills 和约 90 个 MCP 工具是两个维度、schema 包定义知识形态、Minions 在 PGLite 有边界，最后用两个 benchmark 立起了"看懂诚实口径、不混引、成本≠质量"的判断框架。

### 10.2 你现在能做什么

&emsp;&emsp;把能力落到可观察的行为上——下面 6 条，你现在应该都能做到。

&emsp;&emsp;1. 看到一个 brain，你能解释"git markdown 和数据库谁是真相"，能回答"改了源数据库会不会自动更新"，还能说清 PGLite（零配置、嵌入式、单写，适合个人/本地）和 Postgres（多写并发、适合团队/生产）这两种存储引擎各自的适用场景。

&emsp;&emsp;2. 装 GBrain 时，你绝不会用 `npm install -g gbrain`，而是走 git clone + bun link，并用 `--version` 一眼识破抢注包。

&emsp;&emsp;3. 你能把 embedding 开起来，让知识带上语义向量，并用 `doctor` 验证 embed 子项满分。

&emsp;&emsp;4. 你能讲清 HNSW + BM25 + RRF 三路为什么不能直接加分，并解释 k=60 按名次融合的必要性。

&emsp;&emsp;5. 你能用同一句话对照 `search` 和 `think`，并知道 think 输出里最该先看的是末尾的空白分析。

&emsp;&emsp;6. 你能区分 LongMemEval 和 BrainBench 的可信度差异，引用 benchmark 时分清"公开可复现"和"厂商自建"，绝不混引。

### 10.3 三句话自测

&emsp;&emsp;给你三个问题，读完这节课，下面三句你能不能在 30 秒内答出？

&emsp;&emsp;**1. 为什么对大脑跑 `search` 还不够，`think` 补了什么？** —— `search` 只给页面列表，要自己读多页做综合；`think` 替你读完、跨页综合成带引用的答案，末尾还给空白分析。

&emsp;&emsp;**2. RRF 为什么不能把两路分数直接相加？** —— 因为 BM25（tsvector 原始分，量纲偏高）和向量（约 1.0）的分数量纲不可比，绝对值还随语料规模变，直接加会让 BM25 单方面主导；RRF 只看名次（`1/(k+rank)`，k=60）绕开了量纲问题，给两路平等投票权。

&emsp;&emsp;**3. LongMemEval 97.6% 和 BrainBench 49.1% 能放一起比吗？** —— 不能。前者是公开语料 + keyword-only 可复现，后者是厂商自建虚构世界 + LLM judge，两者不在一个坐标系，必须分列、注明来源类型。

---